#Identify DMPs, VMPs and entropy in each dataset and save files for meta-analysis

#### GOLDN ####
path = "/pvol/Preprocessing/Blood/dbGaP/GOLDN"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GOLDN beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GOLDN Phenotypes.txt", sep = "")
rownames(pheno) Pheno$Sample_id
#names(M) <- sub("X","",names(M))
#names(B) <- sub("X","",names(B))
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GOLDN_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GOLDN",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GOLDN_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GOLDN_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("GOLDN_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GOLDN Phenotypes.txt")
rownames(pheno) <- pheno$Sample_id

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#residuals matrix reduced to 148915 CpGs

resid_B <- resid_B[CpGs_to_keep,]

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg02734358",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('GOLDN_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#17968 VMPs with shapiro-wilk

#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(resid_B[cpg,]))
effect <- results_BP$COEF[2]
SE <- results_BP$SE[2]
tiff('GOLDN_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=(meth^2)*100)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% variance methylation at",cpg),
       title = "Squared beta residuals ~ age")
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GOLDN.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")



#### DMPs CTC 

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GOLDN beta.txt")
M <- logit2(B)
pheno <- read.delim("GOLDN Phenotypes.txt") %>%
  mutate(Sample_Name = paste0("X",Sample_Name))
glimpse(pheno)

#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_id <- rownames(celltypes)
pheno$Sample_id <- as.character(pheno$Sample_id)
pheno <- left_join(pheno, celltypes, by = "Sample_id")

write.table(pheno,
            file="GOLDN Phenotypes.txt",
            quote=FALSE,
            row.names=TRUE,
            col.names=TRUE,
            sep='\t')

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GOLDN.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GOLDN_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GOLDN",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GOLDN_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GOLDN_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC 
resid <- data.table::fread("GOLDN_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GOLDN Phenotypes.txt")
rownames(pheno) <- pheno$Sample_id

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#175443 cpgs in reduced cpg matrix

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg13066289",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GOLDN_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
effect <- results_BP$COEF[1]
tiff('GOLDN_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#check the effect size on the beta residuals
pheno_with_meth <- cbind(pheno, meth= as.numeric(resid_B[cpg,]))
tiff('GOLDN_VMPCTCcheck beta squared residuals.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

dev.off()

results_BP %>% filter(PVALUE < 0.005)
#10874 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GOLDN.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### Entropy 

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GOLDN beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- select(B, -V1)
pheno <- read.delim("GOLDN Phenotypes.txt", sep = "")

#run limma and adjust for covariates EXCLUDING age
M <- logit2(B)

design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

which(B_adj == 1)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_id <- rownames(Entropy)
pheno$Sample_id <- as.character(pheno$Sample_id)
pheno <- left_join(pheno, Entropy, by = "Sample_id")

#Save entropy plot
tiff('GOLDN_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.4,0.65)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
#p-value: 5.824e-08
effect = summary(fit)$coefficients[,1][2]
#effect: 0.0002450314
stderr = summary(fit)$coefficients[,2][2]
#stderr = 4.481886e-05

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           NK +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_id <- rownames(Entropy_CTC)
pheno$Sample_id <- as.character(pheno$Sample_id)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_id")

#Save entropy plot
tiff('GOLDN_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.4,0.65)
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
#p-value: 2.118e-05
effectCTC = summary(fitCTC)$coefficients[,1][2]
#effect: 0.0001645221
stderrCTC = summary(fitCTC)$coefficients[,2][2]
#stderr = 3.850116e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.8538517

tiff('GOLDN Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.85",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, 
            "GOLDN Phenotypes.txt")


#### Entropy DMPs 

B <- data.table::fread("GOLDN beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
pheno <- read.delim("GOLDN Phenotypes.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
DMPs <- DMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% DMPs,]
M <- logit2(B)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_id <- rownames(Entropy_DMPs)
pheno$Sample_id <- as.character(pheno$Sample_id)
pheno <- left_join(pheno, Entropy_DMPs, by = "Sample_id")

#Save entropy plot
tiff('GOLDN_Entropy_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy DMPs",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy_DMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#2.888437e-17  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0004564042    
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 5.299575e-05

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_DMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_DMPsCTC <- as.data.frame(Entropy_DMPsCTC)
Entropy_DMPsCTC$Sample_id<- rownames(Entropy_DMPsCTC)
pheno <- left_join(pheno, Entropy_DMPsCTC, by = "Sample_id")

#Save entropy plot
tiff('GOLDN_Entropy_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy DMPs with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_DMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 4.519253e-19     
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002353932    
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =2.616637e-05  

cor.test(pheno$Entropy_DMPs, pheno$Entropy_DMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.61 


write.table(pheno, "GOLDN Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#### Entropy VMPs

B <- data.table::fread("GOLDN beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
pheno <- read.delim("GOLDN Phenotypes.txt", sep = "")

VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
VMPs <- VMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% VMPs,]
M <- logit2(B)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$Sample_id <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "Sample_id")

#Save entropy plot
tiff('GOLDN_Entropy_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy VMPs",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy_VMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#2.381496e-19  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0004819957      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr =5.245699e-05

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_VMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_VMPsCTC <- as.data.frame(Entropy_VMPsCTC)
Entropy_VMPsCTC$Sample_Name<- rownames(Entropy_VMPsCTC)
pheno <- left_join(pheno, Entropy_VMPsCTC, by = "Sample_Name")

#Save entropy plot
tiff('GOLDN_Entropy_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy VMPs with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_VMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 8.057292e-17   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002371093     
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =2.826269e-05   

cor.test(pheno$Entropy_VMPs, pheno$Entropy_VMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.59 


write.table(pheno, "GOLDN Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_CTC, Entropy_DMPs, Entropy_DMPsCTC, Entropy_VMPs, Entropy_VMPsCTC)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GOLDN_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()






#### all DMPs 

path = "/pvol/Preprocessing/Blood"
setwd(path)

DMPs <- read.delim("/pvol/Preprocessing/Blood/Metaanalysis DMPs.txt", sep = "")
DMPs <- DMPs %>% filter(DMPstatus == "DMP")

cpg <- "cg24724428"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))

B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs$MarkerName,]


calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_id <- rownames(Entropy_DMPs)
pheno$Sample_id <- as.character(pheno$Sample_id)
pheno_new <- left_join(pheno, Entropy_DMPs, by = "Sample_id")

ggplot(data = pheno_new, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy DMPs only")+
  theme_minimal()

fit2 <- lm(Entropy_constantVMPs ~ age, pheno_new)
summary(fit2)
pval = summary(fit2)$coefficients[,4][2]
pval
#0.8942637  
effect = summary(fit2)$coefficients[,1][2]
effect
#effect: 5.496088e-06 
stderr = summary(fit2)$coefficients[,2][2]
stderr
#4.134071e-05 



#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GOLDN beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- select(B, -V1)
pheno <- read.delim("GOLDN Phenotypes.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_id <- rownames(Entropy_all)
pheno$Sample_id <- as.character(pheno$Sample_id)
pheno2 <- left_join(pheno, Entropy_all, by = "Sample_id")


ggplot(data = pheno2, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_id <- rownames(Entropy_none)
pheno2 <- left_join(pheno2, Entropy_none, by = "Sample_id")

ggplot(data = pheno2, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


pheno_entropy <- pheno2 %>% select(Sample_id, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_id),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GOLDN_Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#calculate entropy on constant VMPs
VMPs <- VMPs$MarkerName
DMPs <- DMPs$MarkerName
cVMPs <- setdiff(VMPs, DMPs)
B_cVMPs <- B[rownames(B) %in% cVMPs,]

M_cVMPs <- logit2(B_cVMPs)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_cVMPs,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_cVMPs)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_cVMPs)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_cVMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_cVMPs <- as.data.frame(Entropy_cVMPs)
Entropy_cVMPs$Sample_id <- rownames(Entropy_cVMPs)
pheno2 <- left_join(pheno2, Entropy_cVMPs, by = "Sample_id")

ggplot(data = pheno2, aes(x = age, y = Entropy_cVMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GOLDN Entropy constant VMPs",
       ylab = "entropy")+
  theme_minimal()


pheno2 <- pheno2 %>% select(Sample_id, Entropy_all, Entropy_none, Entropy_cVMPs)
pheno$Sample_id <- as.character(pheno$Sample_id)
pheno <- pheno %>% left_join(pheno2, by = "Sample_id")
pheno_entropy <- pheno %>% select(Sample_id, age, Entropy, Entropy_DMPs, Entropy_VMPs, Entropy_all, Entropy_none, Entropy_cVMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_id),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GOLDN_Entropy all age-related CpGs vs non-age-related CpGs, DMPs, VMPs and constant VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 6.111151e-15      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0004044954     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.10186e-05

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5808556      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 2.837769e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.137887e-05 

#summary statistics constant VMPs
fit <- lm(Entropy_cVMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.7352988       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.453253e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.297239e-05





#### WHI ####

path = "/pvol/Preprocessing/Blood/dbGaP/WHI"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("WHI Phenotypes.txt", header = FALSE)
glimpse(pheno)
rownames(pheno) <- pheno$V1
names(pheno)[names(pheno) == "V1"] <- "Sample_Name"
names(pheno)[names(pheno) == "V3"] <- "dbGaP_Sample_ID"
names(pheno)[names(pheno) == "V4"] <- "SAMPLEID"
names(pheno)[names(pheno) == "V5"] <- "SUBJECT_ID"
names(pheno)[names(pheno) == "V6"] <- "BA23_Pull_ID_DNA"
names(pheno)[names(pheno) == "V18"] <- "F2DAYS"
names(pheno)[names(pheno) == "V8"] <- "Tissue"
names(pheno)[names(pheno) == "V19"] <- "age"
names(pheno)[names(pheno) == "V21"] <- "Array"
names(pheno)[names(pheno) == "V22"] <- "Slide"
names(pheno)[names(pheno) == "V23"] <- "Base_Name"
pheno$sex <- "F"

glimpse(pheno)
pheno <- pheno %>% select(Sample_Name,
                          dbGaP_Sample_ID,
                          SUBJECT_ID,
                          Tissue,
                          age,
                          Array,
                          Slide,
                          Base_Name,
                          sex)

write.table(pheno, "WHI Phenotypes.txt",
            col.names = TRUE,
            row.names = TRUE,
            sep = "\t")
pheno$sex <- as.factor(pheno$sex)

#DMPs
design=model.matrix(~age,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('WHI_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "WHI",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('WHI_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="WHI_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("WHI_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("WHI Phenotypes.txt")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] #shapiro wilk reduced matrix to 243156 cpgs
resid_B <- ilogit2(resid) #get the residuals in beta values to obtain regression coefficient and standard error 

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat[cpg,] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('WHI_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('WHI_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=meth*100)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()


setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="WHI.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")



#### DMPs CTC
path = "/pvol/Preprocessing/Blood/dbGaP/WHI"
setwd(path)
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("WHI Phenotypes.txt")

#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno,
            file="WHI Phenotypes.txt",
            quote=FALSE,
            row.names=TRUE,
            col.names=TRUE,
            sep='\t')

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1)

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_WHI.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('WHI_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "WHI",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('WHI_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="WHI_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC
resid <- read.table("WHI_M_res_CTC.txt")
pheno <- read.delim("WHI Phenotypes.txt")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] #reduced cpg matrix to 231108 
resid_B <- ilogit2(resid) #get the residuals in beta values to obtain regression coefficient and standard error 

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg23061578",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('WHI_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('WHI_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth*100)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="WHI.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")




#### Entropy

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
pheno <- read.delim("WHI Phenotypes.txt")

#run limma and adjust for covariets EXCLUDING age
#no covariates for WHI so do not need to run on this dataset

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('WHI_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "WHI Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.3,0.5)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.3131826
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 4.415288e-05
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.376604e-05

#Entropy with CTC 

designCTC = model.matrix(~CD4T +
                           NK +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)
M <- logit2(B)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('WHIN_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "WHI Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.04760497 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 6.503218e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 3.280836e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7557978

tiff('WHI Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.85",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "WHI Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### DMPs vs VMPs entropy 

#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B <- as.data.frame(B) 
B_adj_DMPs <- B[rownames(B) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_Name <- rownames(Entropy_DMPs)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_DMPs, by = "Sample_Name")

tiff('WHI_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "WHI Entropy DMPs only")+
  theme_minimal()
dev.off()

fit <- lm(Entropy_DMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.1895027
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 5.718526e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.356874e-05  

#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B <- as.data.frame(B) 
B_adj_VMPs <- B[rownames(B) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$Sample_Name <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "Sample_Name")

tiff('WHI_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "WHI Entropy VMPs only")+
  theme_minimal()
dev.off()

fit2 <- lm(Entropy_VMPs ~ age, pheno)
summary(fit2)
pval2 = summary(fit2)$coefficients[,4][2]
pval2
#0.3410368  
effect2 = summary(fit2)$coefficients[,1][2]
effect2
#effect:7.699991e-05 
stderr2 = summary(fit2)$coefficients[,2][2]
stderr2
#stderr = 8.085155e-05  

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Condition",
                                                values_to = "Entropy value")


#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|Sample_Name), pheno_entropy)
summary(fit)
# age:ConditionEntropy_VMPs pval 0.418
# ge:ConditionEntropy_DMPs pval 0.748    

library(viridis)
tiff('WHI_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]


#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_all, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- pheno %>% left_join(Entropy_all, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "WHI Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_none, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- pheno %>% left_join(Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "WHI Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('WHI Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2725095                       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 6.980908e-05               
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.360009e-05             

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8377477                         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 7.849941e-06       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.832907e-05         


write.table(pheno, "WHI Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE55763 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE55763"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE55763 beta filtered.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE55763 Phenotypes.txt", row.names = 1)
#glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE55763_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#135646 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE55763",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE55763_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE55763_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs 
resid <- data.table::fread("GSE55763_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE55763 Phenotypes.txt")
rownames(pheno) <- pheno$title

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]

resid_B <- ilogit2(resid)
#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]
fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg21899500",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE55763_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#97893 VMPs with adj pval < 0.005

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE55763.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE55763"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE55763_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE55763 beta filtered.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE55763 Phenotypes.txt")

#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$title <- rownames(celltypes)
pheno$title <- as.character(pheno$title)
pheno <- left_join(pheno, celltypes, by = "title")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$CD4T <- as.numeric(pheno$CD4T)
pheno$Bcell <- as.numeric(pheno$Bcell)
pheno$NK <- as.numeric(pheno$NK)
pheno$Gran <- as.numeric(pheno$Gran)
pheno$Mono <- as.numeric(pheno$CD8T)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
title <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(title = title)
tomerge <- left_join(pheno,
                     tomerge)

write.table(pheno,
            file="/pvol/prelim_data_ARC_2023/GSE55763_pheno.txt",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")


#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE55763.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
#CD4T +
#Bcell +
#CD8T +
#NK +
#Gran,
#pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE55763_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#129738 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE55763",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE55763_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE55763_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")
write.table(pheno,
            file="GSE55763 Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- data.table::fread("GSE55763_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE55763 Phenotypes.txt")
rownames(pheno) <- pheno$title

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg01555614",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE55763_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#72913 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE55763.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE55763"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE55763_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE55763 beta filtered.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
pheno <- read.delim("GSE55763 Phenotypes.txt", row.names = 1)

#Entropy VMPs

VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
VMPs <- VMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% VMPs,]
M <- logit2(B)


design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$title <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "title")

#Save entropy plot
tiff('GSE55763_Entropy_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE55763 Entropy VMPs",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy_VMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#1.49827e-37 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003064835 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 2.356688e-05 

#Entropy with CTC 

designCTC = model.matrix(~ sex + CD4T +
                           NK +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_VMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_VMPsCTC <- as.data.frame(Entropy_VMPsCTC)
Entropy_VMPsCTC$title<- rownames(Entropy_VMPsCTC)
pheno <- left_join(pheno, Entropy_VMPsCTC, by = "title")

#Save entropy plot
tiff('GSE55763_EntropyVMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE55763 VMPs Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_VMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 8.233735e-121  
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002828865   
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 1.149683e-05  

cor.test(pheno$Entropy_VMPs, pheno$Entropy_VMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.585  

tiff('GSE55763 VMPs Entropy ~ Entropy VMPs with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy_VMPs, y = Entropy_VMPsCTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy_VMPs, y = Entropy_VMPsCTC), colour = "black") +
  labs(title = "Pearson's corr = 0.518",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE55763 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

pheno_entropy <- pheno %>% select(title, age, Entropy, Entropy_CTC, Entropy_DMPs, Entropy_DMPsCTC, Entropy_VMPs, Entropy_VMPsCTC)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,title),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE55763_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()

#### Entropy all 

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE55763 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,pheno)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE55763 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


#calculate entropy on constant VMPs
VMPs <- VMPs$MarkerName
DMPs <- DMPs$MarkerName
cVMPs <- setdiff(VMPs, DMPs)
B_cVMPs <- B[rownames(B) %in% cVMPs,]

M_cVMPs <- logit2(B_cVMPs)

design = model.matrix(~sex,pheno)

fit1 <- lmFit(M_cVMPs,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_cVMPs)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_cVMPs)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_cVMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_cVMPs <- as.data.frame(Entropy_cVMPs)
Entropy_cVMPs$Sample_Name <- rownames(Entropy_cVMPs)
pheno <- left_join(pheno, Entropy_cVMPs, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_cVMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE55763 Entropy constant VMPs",
       ylab = "entropy")+
  theme_minimal()



pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_DMPs, Entropy_VMPs, Entropy_all, Entropy_none, Entropy_cVMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE55763_Entropy all age-related CpGs vs non-age-related CpGs, DMPs, VMPs and constant VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 4.769558e-36         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002389088      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 1.878059e-05 

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 2.127234e-07         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -7.009646e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 1.347678e-05  

#summary statistics constant VMPs
fit <- lm(Entropy_cVMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 2.019164e-41        
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0001210949      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 8.83124e-06


write.table(pheno, "GSE55763 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE128235 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE128235"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE128235 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE128235 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

colnames(B) == pheno$title
#DMPs
design=model.matrix(~age +
                      sex +
                      diagnosis,
                    pheno)

#4 samples without a diagnosis removed from analysis 
pheno <- pheno %>% filter(!is.na(diagnosis))
B <- B[,pheno$title]
M <- M[,pheno$title]

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE128235_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#18563 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE128235",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE128235_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE128235_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE128235_M_res.txt")
pheno <- read.delim("GSE128235 Phenotypes.txt")
pheno <- pheno %>% filter(!is.na(diagnosis))

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#296373 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE128235_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#8836 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE128235.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE128235"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE128235_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE128235 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE128235 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$diagnosis <- as.factor(pheno$diagnosis)
pheno$age <- as.numeric(pheno$age)
pheno$CD4T <- as.numeric(pheno$CD4T)
pheno$Bcell <- as.numeric(pheno$Bcell)
pheno$NK <- as.numeric(pheno$NK)
pheno$Gran <- as.numeric(pheno$Gran)
pheno$Mono <- as.numeric(pheno$Mono)

pheno <- pheno %>% filter(!is.na(diagnosis))
B <- B[,pheno$title]
M <- M[,pheno$title]

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
title <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(title = title)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex +
                                            diagnosis,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE128235.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      diagnosis +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE128235_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#18201 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE128235",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE128235_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE128235_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE128235_M_CTC_res.txt")
pheno <- read.delim("GSE128235 Phenotypes.txt")
pheno <- pheno %>% filter(!is.na(diagnosis))

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE128235_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#7165 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE128235.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE128235"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE128235_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE128235 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE128235 Phenotypes.txt")
pheno <- pheno %>% filter(!is.na(diagnosis))

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        diagnosis,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$title <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "title")

#Save entropy plot
tiff('GSE128235_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE128235 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.7846098 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -8.582303e-06 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 3.138467e-05 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           diagnosis +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$title<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "title")

#Save entropy plot
tiff('GSE128235_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE128235 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.00234564   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 9.607677e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 3.142474e-05   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.9881816  

tiff('GSE128235 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.891",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE128235 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


cpg <- "cg13921251"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

png('ConstantVMP GSE55763.png',
    width = 7,
    height = 5,
    unit = 'in',
    res = 600)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.2, colour = "deepskyblue4")+
  labs(y=paste("% methylation at",cpg)) +
  geom_smooth(method = "lm", colour = "slategrey") +
  xlim(19,105) +
  theme_minimal()
dev.off()

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex + diagnosis,
                      pheno)

M_all <- M_all[,pheno$title]
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$title <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "title")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1128235 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

M_none <- M_none[,pheno$title]
fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$title <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "title")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1128235 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(title, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,title),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE1128235 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 6.863647e-05                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001756775          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.377901e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2050711                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -5.496318e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.331919e-05  


write.table(pheno, "GSE1128235 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE99624 ####
path = "/pvol/Preprocessing/Blood/GEO/GSE99624"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GSE99624 Normalized beta values after batch correction.txt")
limma::plotDensities(B, legend = F)
M <- logit2(B)
pheno <- read.delim("GSE99624 Phenotypes.txt",sep = "")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$disease.state <- as.factor(pheno$disease.state)
pheno$Array <- as.factor(pheno$Array)
pheno$Slide <- as.factor(pheno$Slide)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE99624_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#0 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE99624",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE99624_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE99624_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE99624_M_res.txt")
pheno <- read.delim("GSE99624 Phenotypes.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#407859 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00116766",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE99624_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2947 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE99624.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE99624"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE99624_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GSE99624 Normalized beta values after batch correction.txt")
M <- logit2(B)
pheno <- read.delim("GSE99624 Phenotypes.txt",sep = "")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$CD4T <- as.numeric(pheno$CD4T)
pheno$Bcell <- as.numeric(pheno$Bcell)
pheno$NK <- as.numeric(pheno$NK)
pheno$Gran <- as.numeric(pheno$Gran)
pheno$Mono <- as.numeric(pheno$Mono)

#DMPs
design=model.matrix(~age +
                      sex +
                      CD4T +
                      Bcell +
                      Mono +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE99624_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#0 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE99624",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[10]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE99624_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE99624_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE99624_M_CTC_res.txt")
pheno <- read.delim("GSE99624 Phenotypes.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#408575 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg11417025",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE99624_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2950 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE99624.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE99624"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE99624_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy

B <- read.table("GSE99624 Normalized beta values after batch correction.txt")
M <- logit2(B)
pheno <- read.delim("GSE99624 Phenotypes.txt",sep = "")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
pheno <- cbind(pheno, Entropy)

#Save entropy plot
tiff('99624_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "99624 Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.35,0.5)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.2631376
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0001613007  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001423818  

#Entropy with CTC 

designCTC = model.matrix(~ CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
pheno <- cbind(pheno, Entropy_CTC)

#Save entropy plot
tiff('GSE99624_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "99624 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.4,0.47)
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.5391254  
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: -6.371056e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001029646  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  3.932e-09
#cor = 0.7299705 

tiff('99624 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.73",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE99624 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE,
            row.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
pheno <- cbind(pheno, Entropy_all)

ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE99624 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
pheno <- cbind(pheno, Entropy_none)

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE99624 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE99624 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.7233778                     
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -7.799483e-05               
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002190119            

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.3949087                       
effect = summary(fit)$coefficients[,1][2]
effect
#effect:-0.0001160057       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001350801         


write.table(pheno, "GSE99624 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




#### GSE115278 ####
#### 450K

path = "/pvol/Preprocessing/Blood/GEO/GSE115278"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE115278 450K Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE115278 450K Phenotypes.txt", header = FALSE)
glimpse(pheno)

pheno_old <- read_csv("450 IDATS/GSE115278 phenotypes 450.csv")

rownames(pheno) <- pheno$V1
names(pheno)[names(pheno) == "V1"] <- "Sample_Name"
names(pheno)[names(pheno) == "V2"] <- "geo_accession"
names(pheno)[names(pheno) == "V3"] <- "title"
names(pheno)[names(pheno) == "V4"] <- "Sample_Name"
names(pheno)[names(pheno) == "V5"] <- "age"
names(pheno)[names(pheno) == "V6"] <- "Illumina platform"
names(pheno)[names(pheno) == "V7"] <- "sex"
names(pheno)[names(pheno) == "V8"] <- "study"
names(pheno)[names(pheno) == "V9"] <- "waist circumference"
names(pheno)[names(pheno) == "V10"] <- "Array"
names(pheno)[names(pheno) == "V11"] <- "Slide"
names(pheno)[names(pheno) == "V12"] <- "Basename"
names(pheno)[names(pheno) == "V13"] <- "Basename2"
names(pheno)[names(pheno) == "V16"] <- "predictedSex"
names(pheno)[names(pheno) == "V17"] <- "sexmatch"

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

write.table(pheno, "GSE115278 450K Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

#DMPs
design=model.matrix(~age +
                      sex +
                      study,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE115278_450_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#1699 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE115278_450",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
pheno_with_meth <- pheno_with_meth %>% select(-4)
tiff('GSE115278_450_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE115278_450_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- read.table("GSE115278_450_M_res.txt")
pheno <- read.delim("GSE115278 450K Phenotypes.txt", sep = "\t")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#307510 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg12656475",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE115278_450_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#25660 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE115278_450.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE115278"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE115278_450_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()



#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GSE115278 450K Normalized beta values after batch correction.txt")
M <- logit2(B)
pheno <- read.delim("GSE115278 450K Phenotypes.txt")

library(ChAMP)

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)

pheno <- pheno %>% select(-4)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

glimpse(pheno)
write.table(pheno, "GSE115278 450K Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$CD4T <- as.numeric(pheno$CD4T)
pheno$Bcell <- as.numeric(pheno$Bcell)
pheno$NK <- as.numeric(pheno$NK)
pheno$Gran <- as.numeric(pheno$Gran)
pheno$Mono <- as.numeric(pheno$Mono)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            study,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE115278 450K.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      study +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE115278_450_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#6984 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE115278_450",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE115278_450_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE115278_450_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE115278_450_M_CTC_res.txt")
pheno <- read.delim("GSE115278 450K Phenotypes.txt", sep = "\t")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#307653 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg02039400",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE115278_450_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#17458 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE115278_450.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE115278"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE115278_450_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

pheno %>% summarise(mean = mean(age),
                    sd = sd(age), range = range(age))
pheno$sex <- as.factor(pheno$sex)
count(pheno$sex == "M") / nrow(pheno)


#### Entropy

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE115278 450K Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE115278 450K Phenotypes.txt")

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex +
                        study,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE115278_450_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE115278_450 Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.4,0.6)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5078759 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 3.874802e-05
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 5.8459e-05 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           study +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE115278_450_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE115278_450 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.4,0.6)
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.4362365
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect:4.203122e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 4.203122e-05

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.9200265

tiff('GSE115278_450 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9200265",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        study,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE115278 450K Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE115278 450K Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE115278 450K Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2641226              
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 7.819678e-05         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.991343e-05x    

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8314082               
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -1.488824e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.988029e-05     


write.table(pheno, "GSE115278 450K Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### EPIC

path = "/pvol/Preprocessing/Blood/GEO/GSE115278"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE115278 EPIC Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE115278 EPIC Phenotypes.txt",header = FALSE)
glimpse(pheno)

pheno_old <- read_csv("EPIC IDATS/GSE115278 phenotypes EPIC.csv")

rownames(pheno) <- pheno$V1
names(pheno)[names(pheno) == "V1"] <- "Sample_Name"
names(pheno)[names(pheno) == "V4"] <- "geo_accession"
names(pheno)[names(pheno) == "V3"] <- "title"
names(pheno)[names(pheno) == "V5"] <- "age"
names(pheno)[names(pheno) == "V6"] <- "Illumina platform"
names(pheno)[names(pheno) == "V7"] <- "sex"
names(pheno)[names(pheno) == "V8"] <- "study"
names(pheno)[names(pheno) == "V9"] <- "waist circumference"
names(pheno)[names(pheno) == "V10"] <- "Array"
names(pheno)[names(pheno) == "V11"] <- "Slide"
names(pheno)[names(pheno) == "V12"] <- "Basename"
names(pheno)[names(pheno) == "V13"] <- "Basename2"
names(pheno)[names(pheno) == "V16"] <- "predictedSex"
names(pheno)[names(pheno) == "V17"] <- "sexmatch"

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$study <- as.factor(pheno$study)

write.table(pheno, "GSE115278 EPIC Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

#DMPs
design=model.matrix(~age +
                      sex + study,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE115278_EPIC_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#14 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE115278_EPIC",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
pheno_with_meth <- pheno_with_meth %>% select(-4)
tiff('GSE115278_EPIC_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE115278_EPIC_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- read.table("GSE115278_EPIC_M_res.txt")
pheno <- read.delim("GSE115278 EPIC Phenotypes.txt", sep = "\t")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#650628 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg15976217",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE115278_EPIC_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#1824 VMPs  

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE115278_EPIC.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE115278"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE115278_EPIC_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GSE115278 EPIC Normalized beta values after batch correction.txt")
M <- logit2(B)
pheno <- read.delim("GSE115278 EPIC Phenotypes.txt")

library(ChAMP)

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)

pheno <- left_join(pheno, celltypes, by = "Sample_Name")

glimpse(pheno)
write.table(pheno, "GSE115278 EPIC Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$CD4T <- as.numeric(pheno$CD4T)
pheno$CD8T <- as.numeric(pheno$CD8T)
pheno$NK <- as.numeric(pheno$NK)
pheno$Gran <- as.numeric(pheno$Gran)
pheno$Mono <- as.numeric(pheno$Mono)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            study,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE115278 EPIC.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      study +
                      CD4T +
                      CD8T +
                      Mono +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE115278_EPIC_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#29 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE115278_EPIC",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE115278_EPIC_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE115278_EPIC_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE115278_EPIC_M_CTC_res.txt")
pheno <- read.delim("GSE115278 EPIC Phenotypes.txt", sep = "\t")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#653708 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg20517058",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE115278_EPIC_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2099 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE115278_EPIC.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE115278"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE115278_EPIC_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

pheno %>% summarise(mean = mean(age),
                    sd = sd(age), range = range(age))
pheno$sex <- as.factor(pheno$sex)
count(pheno$sex == "M") / nrow(pheno)


#### Entropy 

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE115278 EPIC Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE115278 EPIC Phenotypes.txt", sep = '')

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex +
                        study,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE115278_EPIC_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE115278_EPIC Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.4,0.6)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5765098
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -8.571694e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr =0.0001529961 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           study +
                           CD4T +
                           CD8T +
                           Mono +
                           NK +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE115278_EPIC_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE115278_EPIC Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.4,0.6)
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.9879491 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: -1.935839e-06 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001278587 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.831521

tiff('GSE115278_EPIC Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.831521",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE115278 EPIC Phenotypes.txt")

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        study,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE115278 EPIC Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE115278 EPIC Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE115278 EPIC Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.9066729               
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -2.303176e-05         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001959779     

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.3511263                
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0001635154        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001745811      


write.table(pheno, "GSE115278 EPIC Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE87571 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE87571"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE87571 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE87571 Phenotypes.txt", header = FALSE)
glimpse(pheno)

rownames(pheno) <- pheno$V1
names(pheno)[names(pheno) == "V1"] <- "Sample_Name"
names(pheno)[names(pheno) == "V2"] <- "geo_accession"
names(pheno)[names(pheno) == "V3"] <- "ID"
names(pheno)[names(pheno) == "V4"] <- "age"
names(pheno)[names(pheno) == "V5"] <- "sex"
names(pheno)[names(pheno) == "V6"] <- "Group"
names(pheno)[names(pheno) == "V7"] <- "CD8T"
names(pheno)[names(pheno) == "V8"] <- "CD4T"
names(pheno)[names(pheno) == "V9"] <- "NK"
names(pheno)[names(pheno) == "V10"] <- "Bcell"
names(pheno)[names(pheno) == "V11"] <- "Mono"
names(pheno)[names(pheno) == "V12"] <- "Gran"
names(pheno)[names(pheno) == "V13"] <- "Array"
names(pheno)[names(pheno) == "V14"] <- "Slide"
names(pheno)[names(pheno) == "V15"] <- "Basename"
names(pheno)[names(pheno) == "V16"] <- "Basename2"
names(pheno)[names(pheno) == "V19"] <- "predictedSex"
names(pheno)[names(pheno) == "V20"] <- "sexmatch"

pheno$sex <- as.factor(pheno$sex)
write.table(pheno, "GSE87571 Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE87571_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE87571",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE87571_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE87571_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs
resid <- data.table::fread("GSE87571_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE87571 Phenotypes.txt")
rownames(pheno) <- pheno$Sample_id

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#resid matrix reduced to 287696
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg05282437",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

filter(results_BP, PVALUE < 0.005)
#56096

#Save p-value histogram
tiff('GSE87571_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()


#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE87571_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE87571.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")
 ####

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE87571 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE87571 Phenotypes.txt")

transmute(pheno, sumcelltypes = CD8T + CD4T + Bcell + NK + Mono + Gran)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE87571.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE87571_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE87571",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE87571_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE87571_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC
resid <- data.table::fread("GSE87571_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE87571 Phenotypes.txt")
rownames(pheno) <- pheno$Sample_id

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#285417 cpgs in reduced resid matrix
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16722435",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE87571_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE87571_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=meth*100)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

results_BP %>% filter(PVALUE < 0.005)
#40107 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE87571.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

pheno %>% summarise(mean = mean(age),
                    sd = sd(age),
                    range = range(age))

#### Entropy 
B <- data.table::fread("GSE87571 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE87571 Phenotypes.txt", sep = "")

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE115278_EPIC_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE87571 Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.35,0.48)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 3.211711e-33 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002601156
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr =2.058726e-05 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           NK +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE87571_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE87571 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.35,0.47)
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 6.846388e-19  
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001507627
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 1.651815e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.7728419 

tiff('GSE87571 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7728419 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE87571 Phenotypes.txt")

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design=model.matrix(~ sex,
                    pheno)


fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)

pheno <- pheno %>% mutate(age_group = dplyr::case_when(
  age <= 20            ~ "0-20",
  age > 20 & age <= 40 ~ "21-40",
  age > 40 & age <= 60 ~ "41-60",
  age > 61 & age <= 100  ~ "> 60"
),
# Convert to factor
age_group = factor(
  age_group,
  level = c("0-20", "21-40","41-60", "> 60")
)
)

tiff("GSE87571 density plot age CpGs.tiff",
     height = 6,
     width = 6,
     res = 300,
     units = 'in')
plotDensities(B_adj, 
              group = pheno$age_group, 
              col = c("#A4C6F5","#4FA0E8","#1473D9","#093075"),
              main = "GSE87571 density plot of all age-related CpGs")
dev.off()




#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE87571 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

tiff("GSE87571 density plot non age CpGs.tiff",
     height = 6,
     width = 6,
     res = 300,
     units = 'in')
plotDensities(B_adj, 
              group = pheno$age_group, 
              col = c("#A4C6F5","#4FA0E8","#1473D9","#093075"),
              main = "GSE87571 density plot of non age-related CpGs")
dev.off()

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE87571 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE87571 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 5.431849e-42                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0004573389              
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.155684e-05          

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.4415698                     
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -1.267222e-05       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 1.645813e-05         


write.table(pheno, "GSE87571 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




pheno <- read.delim("GSE87571 Phenotypes.txt")
pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE87571 Entropy all age-related CpGs vs non-age-related CpGs 2.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`, colour = Condition))+ 
  geom_jitter(alpha = 0.3, width = 0.6) +
  geom_smooth(method = "lm")+
  xlim(10,90)+
  ylim(0.1,0.7)+
  scale_colour_manual(values = c("#3F7C85","#747E7E"),
                      name = "", 
                      labels = c("All age-related CpGs","Non age-related CpGs"))+
  scale_alpha(guide = 'none')+
  ggtitle("GSE87571")+
  theme_minimal() +
  theme(plot.title = element_text(size = 12,hjust = 0.5))
dev.off()

#### GSE53740 ####
path = "/pvol/Preprocessing/Blood/GEO/GSE53740"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE53740 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE53740 Phenotypes.txt")

glimpse(pheno)

pheno$Sample_Name == colnames(M)

pheno %>% group_by(race) %>% tally()
M <- M[,!is.na(pheno$race)]
B <- B[,!is.na(pheno$race)]

#DMPs
design=model.matrix(~age +
                      sex +
                      race,
                    pheno)



fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE53740_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE53740",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno %>% filter(!is.na(race)), meth= as.numeric(B[cpg,]))
tiff('GSE53740_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE53740_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("GSE54730_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE54730 Phenotypes.txt")
rownames(pheno) <- pheno$Sample_Name

pheno <- pheno %>% filter(!is.na(race))
design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#residuals matrix reduced to 332662 CpGs

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('GSE53740_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#883 VMPs with shapiro-wilk

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE53740_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("Methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE53740.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

### DMPs with CTC 
path = "/pvol/Preprocessing/Blood/GEO/GSE53740"
setwd(path)

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE53740 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE53740 Phenotypes.txt")
M <- M[,pheno$Sample_Name]
B <- B[,pheno$Sample_Name]

#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")
rownames(pheno) <- pheno$Sample_Name

write.table(pheno,
            file="GSE53740 Phenotypes.txt",
            quote=FALSE,
            row.names=TRUE,
            col.names=TRUE,
            sep='\t')

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            race,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE53740.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      sex +
                      race +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)

#   CD4T +
#   NK +
#   Bcell + 
#   CD8T +
#   Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE53740_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"

create_summary(toptable = results,
               dataset_label = "GSE53740",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE53740_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE53740_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs wtith CTC 

resid <- data.table::fread("GSE53740_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE53740 Phenotypes.txt")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#354134 cpgs in reduced cpg matrix

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE53740_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE53740_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("Methylation at",cpg))
dev.off()


results_BP %>% filter(PVALUE < 0.005)
#775 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE53740.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### Entropy

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE53740 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE53740 Phenotypes.txt", sep = '')

#run limma and adjust for covariets EXCLUDING age

design = model.matrix(~sex +
                        race,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE53740_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53740 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1677049 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001641212  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001184218  

#Entropy with CTC 

designCTC = model.matrix(~race +
                           sex +
                           CD4T +
                           NK +
                           Bcell + 
                           CD8T +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE53740_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53740 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.08266189 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.000105605 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 6.047035e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.515425 

tiff('GSE53740 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.515425",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, 
            "GSE53740 Phenotypes.txt")


#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_Name <- rownames(Entropy_DMPs)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_DMPs, by = "Sample_Name")

tiff('GSE53740_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53740 Entropy DMPs only")+
  theme_minimal()
dev.off()


#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$Sample_Name <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "Sample_Name")

tiff('GSE53740_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53740 Entropy VMPs only")+
  theme_minimal()
dev.off()


pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
pheno_entropy$Condition <- factor(pheno_entropy$Condition, levels = c("Entropy_DMPs", "Entropy_VMPs", "Entropy"))
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|Sample_Name), pheno_entropy)
summary(fit)
# age:ConditionEntropy_VMPs pval 0.0424 *
# age:ConditionEntropy pval 0.8145    

library(viridis)
tiff('GSE53740_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex + race,
                      pheno)

M_all <- M_all[,pheno$Sample_Name]
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53740 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

M_none <- M_none[,pheno$Sample_Name]

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53740 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE53740 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1567571                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002478928            
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001742357        

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5929959                     
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 4.937525e-05          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 9.219107e-05        


write.table(pheno, "GSE53740 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE58045 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE58045"
setwd(path)

library(GEOquery)
library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### correcting for cell type in 27K 

beta <- read.delim("GSE58045_beta.txt", sep = " ", header = T, row.names = 1)
pheno <- read.delim("GSE58045 Phenotypes.txt", header = T, sep = " ")
rownames(pheno) <- pheno$ID

memory.limit(size = 1000000)
champ_celltypes <- champ.refbase(beta=beta,
                                 arraytype="27K")

cell_types <- champ_celltypes[[2]]

pheno <- cbind(pheno, cell_types)

write.table(pheno, "GSE58045 Phenotypes.txt")

beta_corrected <- champ_celltypes[[1]]
write.table(beta_corrected, "beta_celltypecorrected_champ.txt")

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}
beta <- read.delim("GSE58045 beta after filtering.txt", header = T, row.names = 1, sep = "")
pheno <- read.delim("GSE58045 Phenotypes.txt")

glimpse(pheno)
pheno <- pheno %>% modify_at(c(1,3,4,5,6), as.factor)

plotDensities(beta, legend = F)

M <- logit2(beta)
plotDensities(M, legend = F)

design <- model.matrix(~ age, pheno)

pheno$family_number <- as.factor(pheno$family_number)

#paired design, twins at random effect

corfit <- duplicateCorrelation(M,
                               design,
                               block = pheno$family_number)
cor <- corfit$consensus
cor
#[1] 0.2683226

fit1_M <- lmFit(object = M,
                design = design, 
                block = pheno$family_number,
                correlation = cor)

fit2_M <- eBayes(fit1_M)

corfit_B <- duplicateCorrelation(B,
                                 design,
                                 block = pheno$family_number)
corB <- corfit_B$consensus
corB
#[1] 0.2497746

fit1_B <- lmFit(object = B,
                design = design, 
                block = pheno$family_number,
                correlation = corB)

fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE58045_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE58045",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE58045_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE58045_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs
resid <- data.table::fread("GSE58045_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE58045 Phenotypes.txt")
rownames(pheno)

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#residuals matrix reduced to 18616 CpGs

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('GSE58045_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#89 VMPs with shapiro-wilk

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE58045_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=(meth^2)*100)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE58045.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GSE58045 beta after filtering.txt")
M <- logit2(B)
pheno <- read.delim("GSE58045 Phenotypes.txt")


#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      CD4T +
                      CD8T +
                      Bcell + 
                      Mono +
                      Gran,
                    pheno)

corfit <- duplicateCorrelation(M,
                               design,
                               block = pheno$family_number)
cor <- corfit$consensus
cor
#[1] 0.2512177

fit1_M <- lmFit(object = M,
                design = design, 
                block = pheno$family_number,
                correlation = cor)

fit2_M <- eBayes(fit1_M)

corfit_B <- duplicateCorrelation(B,
                                 design,
                                 block = pheno$family_number)
corB <- corfit_B$consensus
corB
#[1] 0.2356784

fit1_B <- lmFit(object = B,
                design = design, 
                block = pheno$family_number,
                correlation = corB)

fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE58045_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE58045",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE58045_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE58045_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMP analysis CTC
resid <- data.table::fread("GSE58045_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE58045 Phenotypes.txt")
rownames(pheno) <- pheno$Sample_id

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#19664 cpgs in reduced cpg matrix

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE58045_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

tiff('GSE58045_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()


setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE58045.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### Entropy 

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE58045 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- select(B, -V1)
pheno <- read.delim("GSE58045 Phenotypes.txt", sep = "")

#run limma and adjust for covariets EXCLUDING age
M <- logit2(B)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$ID <- rownames(Entropy)
pheno$Sample_id <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy, by = "ID")

#Save entropy plot
tiff('GSE58045_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE58045 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.001603156 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0007853499 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0002448917 

#Entropy with CTC 

designCTC = model.matrix(~ CD4T +
                           CD8T +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)

corfit <- duplicateCorrelation(M,
                               designCTC,
                               block = pheno$family_number)
cor <- corfit$consensus
cor
#[1] 0.2535736

fit1_M <- lmFit(object = M,
                design = design, 
                block = pheno$family_number,
                correlation = cor)

fit2_M <- eBayes(fit1_M)

resid_M_CTC <- residuals(fit2_M, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$ID <- rownames(Entropy_CTC)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_CTC, by = "ID")

#Save entropy plot
tiff('GSE58045_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE58045 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal() 
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.9427179 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: -1.474148e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
effectCTC
#stderr = -1.474148e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.8192872 

tiff('GSE58045 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.82",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, 
            "GSE58045 Phenotypes.txt")

#### DMPs vs VMPs entropy 

#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj_DMPs <- B[rownames(B) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$ID <- rownames(Entropy_DMPs)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_DMPs, by = "ID")

tiff('GSE58045_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE58045 Entropy DMPs only")+
  theme_minimal()
dev.off()


#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj_VMPs <- B[rownames(B) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$ID <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "ID")

tiff('GSE58045_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE58045 Entropy VMPs only")+
  theme_minimal()
dev.off()


pheno_entropy <- pheno %>% select(ID, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
pheno_entropy$Condition <- factor(pheno_entropy$Condition, levels = c("Entropy_DMPs", "Entropy_VMPs", "Entropy"))
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|ID), pheno_entropy)
summary(fit)
# age:ConditionEntropy_VMPs pval 0.07710 
# ge:ConditionEntropy_DMPs pval 0.70635    

library(viridis)
tiff('GSE58045_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_all, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$ID <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "ID")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE58045 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]


#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_none, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$ID <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "ID")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE58045 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(ID, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE58045 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 7.352298e-05                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.001002084              
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002465622          

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.05238299                       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0005684551            
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002909698          


write.table(pheno, "GSE58045 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### NAS ####

path = "/pvol/Preprocessing/Blood/dbGaP/NAS"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("NAS beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("NAS Phenotypes.txt")


pheno$id <- as.factor(pheno$id)
pheno$year <- as.factor(pheno$year)

pheno %>% group_by(id) %>% summarise(n = n()) %>% filter(n==1)
pheno %>% group_by(year) %>% summarise(n = n())

row.has.inf <- apply(M,1,function(x){any(is.infinite(x))})
M <- M[!row.has.inf,]


#DMPs

design=model.matrix(~age + year,
                    pheno)

#cannot run duplicate correlation on the entire dataset because it takes too long to run. 
#Randomly select a subset of rows and run duplicate correlation on those samples 

Msubset <- M[sample(nrow(M),size=20000,replace=FALSE),]

corfit <- duplicateCorrelation(Msubset, #we fit limma models on M values and not beta values
                               design,
                               block=pheno$id) 
corfit$consensus
#0.2839511

tomerge <- tibble(Sample_Name = colnames(M))
pheno <- left_join(tomerge, pheno)


fit1_M <- lmFit(M,
                design,
                block = pheno$id,
                correlation = 0.2839511)


fit2_M <- eBayes(fit1_M)

B <- B[!row.has.inf,]

Bsubset <- B[sample(nrow(B),size=1000,replace=FALSE),]
corfitB <- duplicateCorrelation(Bsubset, #we fit limma models on M values and not beta values
                                design,
                                block=pheno$id) 
corfitB$consensus
#0.2686253 #100 samples
#0.3238644 #500 samples
#0.332095 #1000 samples

fit1_B <- lmFit(B,
                design,
                block = pheno$id,
                correlation = 0.332095)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('NAS_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "NAS",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('NAS_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="
            _M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

write.table(B, 
            "NAS beta normalised and batch corrected.txt",
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

#### VMPs

resid <- data.table::fread("NAS_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("NAS Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
#residuals matrix reduced to 105902 CpGs

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg02734358",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('NAS_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#62479 VMPs without shapiro-wilk

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('NAS_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("%  methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="NAS.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")
#### DMPs CTC 

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("NAS beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("NAS Phenotypes.txt")


pheno$id <- as.factor(pheno$id)
pheno$year <- as.factor(pheno$year)

#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_id <- rownames(celltypes)
pheno$Sample_id <- as.character(pheno$Sample_id)
pheno <- left_join(pheno, celltypes, by = "Sample_id")

write.table(pheno,
            file="NAS Phenotypes.txt",
            quote=FALSE,
            row.names=TRUE,
            col.names=TRUE,
            sep='\t')

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ year,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_NAS.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      year +
                      # CD4T +
                      # Bcell +
                      # CD8T +
                      # NK +
                      # Gran,
                      # pheno)
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)

row.has.inf <- apply(M,1,function(x){any(is.infinite(x))})
M <- M[!row.has.inf,]

Msubset <- M[sample(nrow(M),size=500,replace=FALSE),]

corfit <- duplicateCorrelation(Msubset, #we fit limma models on M values and not beta values
                               design,
                               block=pheno$id) 
corfit$consensus.correlation
#0.2547834

tomerge2 <- tibble(Sample_Name = colnames(M))
pheno <- left_join(tomerge2, tomerge)


fit1_M <- lmFit(M,
                design,
                block = pheno$id,
                correlation = 0.2547834)


fit2_M <- eBayes(fit1_M)

B <- B[!row.has.inf,]

Bsubset <- B[sample(nrow(B),size=500,replace=FALSE),]
corfitB <- duplicateCorrelation(Bsubset, #we fit limma models on M values and not beta values
                                design,
                                block=pheno$id) 
corfitB$consensus
#0.2770503

fit1_B <- lmFit(B,
                design,
                block = pheno$id,
                correlation = 0.2770503)
fit2_B <- eBayes(fit1_B)


coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('NAS_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "NAS",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('NAS_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="NAS_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC
resid <- data.table::fread("NAS_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("NAS Phenotypes.txt")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
#88546 cpgs in reduced cpg matrix

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg13066289",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('NAS_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
effect <- results_BP$COEF[1]
tiff('NAS_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()


results_BP %>% filter(PVALUE < 0.005)
#47437 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="NAS.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

####Entropy 

B <- data.table::fread("NAS beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("NAS Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~year,
                      pheno)
row.has.inf <- apply(M,1,function(x){any(is.infinite(x))})
M <- M[!row.has.inf,]

fit1 <- lmFit(M,
              design,
              block = pheno$id,
              correlation = 0.2839511)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('NAS_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "NAS Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.4407512 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -4.778734e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 6.196993e-05 

#Entropy with CTC 

designCTC = model.matrix(~year +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  design,
                  block = pheno$id,
                  correlation = 0.2547834)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('NAS_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "NAS Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.2034583   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 7.786162e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 6.119591e-05   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.9893324  

tiff('NAS Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9893324",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "NAS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~year,
                      pheno)
row.has.inf <- apply(M_all,1,function(x){any(is.infinite(x))})
M_all <- M_all[!row.has.inf,]

fit1 <- lmFit(M_all,
              design,
              block = pheno$id,
              correlation = 0.2839511)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)


#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- pheno %>% left_join(Entropy_all, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "NAS Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

row.has.inf <- apply(M_none,1,function(x){any(is.infinite(x))})
M_none <- M_none[!row.has.inf,]

fit1 <- lmFit(M_none,
              design,
              block = pheno$id,
              correlation = 0.2839511)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)
#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- pheno %>% left_join(Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "NAS Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('NAS Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.04939602                      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001585399               
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 8.060789e-05             

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5675632                        
effect = summary(fit)$coefficients[,1][2]
effect
#effect:-3.340374e-05       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.842116e-05         


write.table(pheno, "NAS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### BIOS ####
path = "/pvol/Preprocessing/Blood/BIOS"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("BIOS Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("BIOS Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

#DMPs
design=model.matrix(~age +
                      predictedSex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('BIOS_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#124439 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "BIOS",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="BIOS_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('BIOS_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs
resid <- read.table("BIOS_M_res.txt")
pheno <- read.delim("BIOS Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#296373 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('BIOS_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#52337 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="BIOS.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/BIOS"
setwd(path)
#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('BIOS_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg)) +
  ylim(0.45,0.8)
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("BIOS Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("BIOS Phenotypes.txt")
glimpse(pheno)

library(ChAMP)
celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

pheno %>% transmute(sumcelltypes = Bcell + CD4T + CD8T + Gran + Mono + NK)

write.table(pheno,
            "BIOS Phenotypes.txt")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ predictedSex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_BIOS.rds')

#DMPs
design=model.matrix(~age +
                      predictedSex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)


coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('BIOS_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#122995 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"

create_summary(toptable = results,
               dataset_label = "BIOS",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('BIOS_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="BIOS_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- read.table("BIOS_M_CTC_res.txt")
pheno <- read.delim("BIOS Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('BIOS_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#35374 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="BIOS.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/BIOS"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('BIOS_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy

B <- data.table::fread("BIOS Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("BIOS Phenotypes.txt", sep = "")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~predictedSex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('BIOS_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#5.624614e-12 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001088927  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 1.567108e-05 

#Entropy with CTC 

designCTC = model.matrix(~predictedSex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('BIOS_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 1.667965e-15   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001098711  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 1.363851e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8784908  

tiff('BIOS Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8784908",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "BIOS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


cpg <- "cg13921251"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

png('ConstantVMP GSE55763.png',
    width = 7,
    height = 5,
    unit = 'in',
    res = 600)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.2, colour = "deepskyblue4")+
  labs(y=paste("% methylation at",cpg)) +
  geom_smooth(method = "lm", colour = "slategrey") +
  xlim(19,105) +
  theme_minimal()
dev.off()

#### Entropy DMPs

B <- data.table::fread("BIOS Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)

pheno <- read.delim("BIOS Phenotypes.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
DMPs <- DMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% DMPs,]
M <- logit2(B)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~predictedSex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_Name <- rownames(Entropy_DMPs)
pheno <- left_join(pheno, Entropy_DMPs, by = "Sample_Name")

#Save entropy plot
tiff('BIOS_Entropy_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy DMPs",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy_DMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#5.827059e-34 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002729069   
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 2.187291e-05 

#Entropy with CTC 

designCTC = model.matrix(~predictedSex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_DMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_DMPsCTC <- as.data.frame(Entropy_DMPsCTC)
Entropy_DMPsCTC$Sample_Name<- rownames(Entropy_DMPsCTC)
pheno <- left_join(pheno, Entropy_DMPsCTC, by = "Sample_Name")

#Save entropy plot
tiff('BIOS_Entropy_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy DMPs with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_DMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 3.016713e-59    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002730863   
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =1.60396e-05  

cor.test(pheno$Entropy_DMPs, pheno$Entropy_DMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.77 


write.table(pheno, "BIOS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#### Entropy VMPs

B <- data.table::fread("BIOS Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)

pheno <- read.delim("BIOS Phenotypes.txt", sep = "")

VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
VMPs <- VMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% VMPs,]
M <- logit2(B)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~predictedSex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$Sample_Name <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "Sample_Name")

#Save entropy plot
tiff('BIOS_Entropy_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy VMPs",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy_VMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#2.632533e-34 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003053918    
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 2.433968e-05 

#Entropy with CTC 

designCTC = model.matrix(~predictedSex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_VMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_VMPsCTC <- as.data.frame(Entropy_VMPsCTC)
Entropy_VMPsCTC$Sample_Name<- rownames(Entropy_VMPsCTC)
pheno <- left_join(pheno, Entropy_VMPsCTC, by = "Sample_Name")

#Save entropy plot
tiff('BIOS_Entropy_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy VMPs with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_VMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 6.415341e-64    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0003075374    
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =0.0003075374   

cor.test(pheno$Entropy_VMPs, pheno$Entropy_VMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.77 


write.table(pheno, "BIOS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_CTC, Entropy_DMPs, Entropy_DMPsCTC, Entropy_VMPs, Entropy_VMPsCTC)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('BIOS_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#### Entropy all 
B <- data.table::fread("BIOS Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)

pheno <- read.delim("BIOS Phenotypes.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~predictedSex,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~predictedSex,
                      pheno)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on constant VMPs
VMPs <- VMPs$MarkerName
DMPs <- DMPs$MarkerName
cVMPs <- setdiff(VMPs, DMPs)
B_cVMPs <- B[rownames(B) %in% cVMPs,]

M_cVMPs <- logit2(B_cVMPs)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~predictedSex,
                      pheno)

fit1 <- lmFit(M_cVMPs,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_cVMPs)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_cVMPs)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_cVMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_cVMPs <- as.data.frame(Entropy_cVMPs)
Entropy_cVMPs$Sample_Name <- rownames(Entropy_cVMPs)
pheno <- left_join(pheno, Entropy_cVMPs, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_cVMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "BIOS Entropy constant VMPs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_DMPs, Entropy_VMPs, Entropy_all, Entropy_none, Entropy_cVMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('BIOS_Entropy all age-related CpGs vs non-age-related CpGs, DMPs, VMPs and constant VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()

#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 2.484315e-28       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002303725      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 2.041515e-05

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.0005125831       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -5.696658e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 1.635944e-05 

#summary statistics constant VMPs
fit <- lm(Entropy_cVMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 1.181521e-12       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -8.998213e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 1.254433e-05


write.table(pheno, "BIOS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#### GSE77445 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE77445"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE77445 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE77445 Phenotypes.txt")

ChAMP::champ.SVD(beta = B,
                 pd = pheno %>% select(age,
                                       Cortisol.stress.response.AUC,
                                       sex,
                                       position.ch1,
                                       race.ch1,
                                       sentrix.ch1),
                 resultsDir = "./CHAMP_SVDimages")

rownames(pheno) <- pheno$ID

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE77445_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE77445",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE77445_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE77445_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("GSE54730_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE54730 Phenotypes.txt")
rownames(pheno) <- pheno$ID

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#residuals matrix reduced to 339089 CpGs

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg04789818",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('GSE77445_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#3128 VMPs with shapiro-wilk

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

tiff('GSE77445_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE77445.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

### DMPs with CTC 

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE77445 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE77445 Phenotypes.txt")

#use champ package to obtain cell type estimates 

transmute(pheno, sumcelltypes = CD4T + CD8T + Bcell + Mono + NK + Gran)


#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE77445.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      sex +
                      CD4T +
                      NK +
                      Bcell + 
                      CD8T +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE77445_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE77445",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE77445_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE77445_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs wtith CTC

resid <- data.table::fread("GSE77445_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE77445 Phenotypes.txt")
rownames(pheno) <- pheno$ID

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#339199 cpgs in reduced cpg matrix

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg14066865",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE77445_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
effect <- results_BP$COEF[1]
tiff('GSE77445_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()


results_BP %>% filter(PVALUE < 0.005)
#2200 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE77445.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### Entropy 

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE77445 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE77445 Phenotypes.txt", sep = " ")

#run limma and adjust for covariets EXCLUDING age

design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$ID <- rownames(Entropy)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy, by = "ID")

#Save entropy plot
tiff('GSE77445_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE77445 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value:0.0009426799
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002774031
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 8.087599e-05  

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           NK +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$ID <- rownames(Entropy_CTC)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_CTC, by = "ID")

#Save entropy plot
tiff('GSE77445_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE77445 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.0006969362
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.000258582 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 7.340072e-05   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.9135549 

tiff('GSE77445 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9135549",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, 
            "GSE77445 Phenotypes.txt")


#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$ID <- rownames(Entropy_DMPs)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_DMPs, by = "ID")

tiff('GSE77445_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE77445 Entropy DMPs only")+
  theme_minimal()
dev.off()


#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$ID <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "ID")

tiff('GSE77445_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE77445 Entropy VMPs only")+
  theme_minimal()
dev.off()


pheno_entropy <- pheno %>% select(ID, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Condition",
                                                values_to = "Entropy value")

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
pheno_entropy$Condition <- factor(pheno_entropy$Condition, levels = c("Entropy_DMPs", "Entropy_VMPs", "Entropy"))
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|ID), pheno_entropy)
summary(fit)
# age:ConditionEntropy_VMPs pval 0.000795 ***
# age:ConditionEntropy pval 0.997625        

library(viridis)
tiff('GSE77445_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design=model.matrix(~ sex,
                    pheno)


fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$ID <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "ID")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE77445 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$ID <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "ID")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE77445 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(ID, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE77445 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.0001002756                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003674748              
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 8.990476e-05         

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.08355632                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001541596        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 8.80165e-05         


write.table(pheno, "GSE77445 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)





#### JHS ####

path = "/pvol/Preprocessing/Blood/JHS"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("JHS beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("JHS Phenotypes.txt")
glimpse(pheno)

#do multiple imputations on phenotypes 
gc()

library(mice)


md.pattern(pheno)
glimpse(pheno)

pheno$edu3cat <- as.factor(pheno$edu3cat)
pheno$Income <- as.factor(pheno$Income)
pheno$occupation <- as.fatocr(pheno$occupation)
pheno$fmlyinc <- as.factor(pheno$fmlyinc)
pheno$currentSmoker <- as.factor(pheno$currentSmoker)
pheno$alc <- as.factor(pheno$alc)
pheno$strokeHx <- as.factor(pheno$strokeHx)
pheno$CHDHx <- as.factor(pheno$CVDHx)
pheno$Diabetes <- as.factor(pheno$Diabetes)
pheno$HTN <- as.factor(pheno$HTN)
pheno$DMmeds <- as.factor(pheno$DMmeds)
pheno$BPmedsSelf <- as.factor(pheno$BPmedsSelf)


pheno_imputed <- pheno %>% select(-c("Sample_Name",
                                     "Sample_Group",
                                     "Sentrix_ID",
                                     "Sentrix_Position",
                                     "Sample_Column",
                                     "visit",
                                     "VisitDate"))
imputed <- mice(data = pheno_imputed,
                m = 10)

imputed$imp

pheno_imputed <- complete(imputed)
predmat <- imputed$predictorMatrix

plot(imputed)

pheno_withimputations <- pheno %>% select("Sample_Name", "Sentrix_ID","Sentrix_Position") %>% cbind(pheno_imputed)

write.table(pheno_withimputations,
            "JHS Phenotypes with imputations.txt")

pheno_withimputations$Sample_Name == colnames(B)
#DMPs
design=model.matrix(~age +
                      sex +
                      edu3cat +
                      BMI +
                      currentSmoker +
                      alc,
                    pheno_withimputations)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('JHS_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#177219 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "JHS",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="JHS_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno_withimputations, meth= as.numeric(B[cpg,]))
tiff('JHS_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("JHS_M_res.txt")
pheno <- read.delim("JHS Phenotypes with imputations.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,] 
#254602 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

results_metaRE <- tibble(CPG = rownames(results),
                         TESTSTAT = results$TestStat,
                         N = rep(ncol(w),nrow(w)),
                         COEF = results$logFC,
                         SE = results$SE,
                         PVALUE = results$chisq_pval,
                         BP = ncol(resid)*fitted_sqrd/rowSums(w^2))
#Save p-value histogram
tiff('JHS_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#126921 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="JHS.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")
setwd("/pvol/EWAS/Blood/RE meta")
write.table(results_metaRE,
            file="JHS_VMPs.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/JHS"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('JHS_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("JHS beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("JHS Phenotypes with imputations.txt", sep = " ")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "JHS Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            edu3cat +
                                            BMI +
                                            currentSmoker +
                                            alc,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_JHS.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      edu3cat +
                      BMI +
                      currentSmoker +
                      alc +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('JHS_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#174297 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "JHS",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('JHS_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="JHS_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- data.table::fread("JHS_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("JHS Phenotypes with imputations.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)
gc()

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)
gc()

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('JHS_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#100637 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="JHS.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/JHS"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('JHS_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("JHS beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("JHS Phenotypes.txt")
pheno2 <- read.delim("JHS Phenotypes with imputations.txt", sep = " ")

#run limma and adjust for covariates EXCLUDING age

design = model.matrix(~sex +
                        edu3cat +
                        BMI +
                        currentSmoker +
                        alc,
                      pheno2)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('JHS_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.00319957    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 8.835696e-05   
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 2.993141e-05 


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           edu3cat +
                           BMI +
                           currentSmoker +
                           alc +
                           CD4T +
                           CD8T +
                           Mono +
                           Bcell +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('JHS_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 4.047926e-05   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 9.463708e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 2.299669e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 1   

tiff('JHS Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 1 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "JHS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#### Entropy DMPs 

B <- data.table::fread("JHS beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
pheno <- read.delim("JHS Phenotypes.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
DMPs <- DMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% DMPs,]
M <- logit2(B)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_Name <- rownames(Entropy_DMPs)
pheno <- left_join(pheno, Entropy_DMPs, by = "Sample_Name")

#Save entropy plot
tiff('JHS_Entropy_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy DMPs",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy_DMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#4.022001e-06 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001805276     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 3.903314e-05

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_DMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_DMPsCTC <- as.data.frame(Entropy_DMPsCTC)
Entropy_DMPsCTC$Sample_Name<- rownames(Entropy_DMPsCTC)
pheno <- left_join(pheno, Entropy_DMPsCTC, by = "Sample_Name")

#Save entropy plot
tiff('JHS_Entropy_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy DMPs with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_DMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 4.519253e-19     
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002353932    
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =2.616637e-05  

cor.test(pheno$Entropy_DMPs, pheno$Entropy_DMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.61 


write.table(pheno, "JHS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#### Entropy VMPs 

B <- data.table::fread("JHS beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
pheno <- read.delim("JHS Phenotypes.txt", sep = "")

VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
VMPs <- VMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% VMPs,]
M <- logit2(B)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        edu3cat +
                        BMI +
                        currentSmoker +
                        alc,
                      pheno2)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$Sample_Name <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "Sample_Name")

#Save entropy plot
tiff('JHS_Entropy_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy VMPs",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy_VMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#1.896085e-05  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001851767      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.317845e-05  

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_VMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_VMPsCTC <- as.data.frame(Entropy_VMPsCTC)
Entropy_VMPsCTC$Sample_Name<- rownames(Entropy_VMPsCTC)
pheno <- left_join(pheno, Entropy_VMPsCTC, by = "Sample_Name")

#Save entropy plot
tiff('JHS_Entropy_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy VMPs with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_VMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 8.057292e-17   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002371093     
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =2.826269e-05   

cor.test(pheno$Entropy_VMPs, pheno$Entropy_VMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.59 


write.table(pheno, "JHS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_CTC, Entropy_DMPs, Entropy_DMPsCTC, Entropy_VMPs, Entropy_VMPsCTC)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('JHS_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()

####

B <- data.table::fread("JHS beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- select(B, -V1)
pheno <- read.delim("JHS Phenotypes.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        edu3cat +
                        BMI +
                        currentSmoker +
                        alc,
                      pheno2)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        edu3cat +
                        BMI +
                        currentSmoker +
                        alc,
                      pheno2)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('JHS_Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#calculate entropy on constant VMPs
VMPs <- VMPs$MarkerName
DMPs <- DMPs$MarkerName
cVMPs <- setdiff(VMPs, DMPs)
B_cVMPs <- B[rownames(B) %in% cVMPs,]

M_cVMPs <- logit2(B_cVMPs)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        edu3cat +
                        BMI +
                        currentSmoker +
                        alc,
                      pheno2)

fit1 <- lmFit(M_cVMPs,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_cVMPs)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_cVMPs)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_cVMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_cVMPs <- as.data.frame(Entropy_cVMPs)
Entropy_cVMPs$Sample_Name <- rownames(Entropy_cVMPs)
pheno <- left_join(pheno, Entropy_cVMPs, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_cVMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "JHS Entropy constant VMPs",
       ylab = "entropy")+
  theme_minimal()



pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_DMPs, Entropy_VMPs, Entropy_all, Entropy_none, Entropy_cVMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('JHS_Entropy all age-related CpGs vs non-age-related CpGs, DMPs, VMPs and constant VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 4.558487e-05        
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001525969     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.733259e-05 

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 2.127234e-07         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -7.009646e-05      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 1.347678e-05  

#summary statistics constant VMPs
fit <- lm(Entropy_cVMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 2.019164e-41         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0001210949      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 8.83124e-06


write.table(pheno, "JHS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




#### FHS ####

path = "/pvol/Preprocessing/Blood/dbGaP/FHS"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("FHS beta values after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("FHS Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('FHS_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "FHS",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('FHS_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="FHS_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")
#### VMPs 

resid <- data.table::fread("FHS_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("FHS Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#resid matrix reduced ro 176767 cpgs

resid_B <- minfi::ilogit2(resid)

library(limma)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg00594952",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('FHS_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#35996 VMPs with pval < 0.005 without shaipro-wilk 

#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
effect <- results_BP$COEF[2]
tiff('FHS_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg),
       title = paste("effect at", effect))
dev.off()

pheno %>% summarise(mean = mean(age),
                    sd = sd(age),
                    range = range(age))

pheno %>% group_by(sex) %>% tally()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="FHS.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### DMPs with CTC 

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("FHS beta values after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("FHS Phenotypes.txt")
#glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")

celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)

pheno <- left_join(pheno, celltypes, by = "Sample_Name")
write.table(pheno, "FHS Phenotypes.txt",
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_FHS.rds')


#DMPs
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('FHS_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "FHS",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('FHS_DMPCTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="FHS_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")
#### VMPs with CTC

resid <- data.table::fread("FHS_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("FHS Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#resid matrix reduced to 172749 cpgs

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg21764262",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('FHS_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#25995 VMPs with pval < 0.005 without shaipro-wilk 

#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
effect <- results_BP$COEF[2]
tiff('FHS_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg),
       title = paste("effect at", effect))
dev.off()

pheno %>% summarise(mean = mean(age),
                    sd = sd(age),
                    range = range(age))

pheno %>% group_by(sex) %>% tally()

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="FHS.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")


cpg <- "cg15007959"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))


#### Entropy 
B <- data.table::fread("FHS beta values after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("FHS Phenotypes.txt")

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('FHS Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.4,0.53)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1877697  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 3.869656e-05
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr =2.936972e-05 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           Gran +
                           NK,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('FHS_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 2.780415e-05  
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 9.453269e-05 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =2.251557e-05

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.7713402

tiff('FHS Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7713402 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "FHS Phenotypes.txt")

#### Entropy DMPs

B <- data.table::fread("FHS beta values after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
pheno <- read.delim("FHS Phenotypes.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
DMPs <- DMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% DMPs,]
M <- logit2(B)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_Name <- rownames(Entropy_DMPs)
pheno <- left_join(pheno, Entropy_DMPs, by = "Sample_Name")

#Save entropy plot
tiff('FHS_Entropy_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy DMPs",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy_DMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.002956028 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001307468    
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.394547e-05

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_DMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_DMPsCTC <- as.data.frame(Entropy_DMPsCTC)
Entropy_DMPsCTC$Sample_Name<- rownames(Entropy_DMPsCTC)
pheno <- left_join(pheno, Entropy_DMPsCTC, by = "Sample_Name")

#Save entropy plot
tiff('FHS_Entropy_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy DMPs with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_DMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 4.519253e-19     
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002353932    
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =2.616637e-05  

cor.test(pheno$Entropy_DMPs, pheno$Entropy_DMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.61 


write.table(pheno, "FHS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#### Entropy VMPs

B <- data.table::fread("FHS beta values after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
pheno <- read.delim("FHS Phenotypes.txt", sep = "")

VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
VMPs <- VMPs$MarkerName
B <- as.data.frame(B) 
B <- B[rownames(B) %in% VMPs,]
M <- logit2(B)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$Sample_Name <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "Sample_Name")

#Save entropy plot
tiff('FHS_Entropy_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy VMPs",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy_VMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.02482554  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001102268     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.908863e-05 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_VMPsCTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_VMPsCTC <- as.data.frame(Entropy_VMPsCTC)
Entropy_VMPsCTC$Sample_Name<- rownames(Entropy_VMPsCTC)
pheno <- left_join(pheno, Entropy_VMPsCTC, by = "Sample_Name")

#Save entropy plot
tiff('FHS_Entropy_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPsCTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy VMPs with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_VMPsCTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 8.057292e-17   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002371093     
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr =2.826269e-05   

cor.test(pheno$Entropy_VMPs, pheno$Entropy_VMPsCTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.59 


write.table(pheno, "FHS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_CTC, Entropy_DMPs, Entropy_DMPsCTC, Entropy_VMPs, Entropy_VMPsCTC)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('FHS_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#### all DMPs

B <- data.table::fread("FHS beta values after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- select(B, -V1)
pheno <- read.delim("FHS Phenotypes.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('FHS_Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#calculate entropy on constant VMPs
VMPs <- VMPs$MarkerName
DMPs <- DMPs$MarkerName
cVMPs <- setdiff(VMPs, DMPs)
B_cVMPs <- B[rownames(B) %in% cVMPs,]

M_cVMPs <- logit2(B_cVMPs)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_cVMPs,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_cVMPs)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_cVMPs)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_cVMPs <- apply(B_adj, 2, calculate_entropy)
Entropy_cVMPs <- as.data.frame(Entropy_cVMPs)
Entropy_cVMPs$Sample_Name <- rownames(Entropy_cVMPs)
pheno <- left_join(pheno, Entropy_cVMPs, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_cVMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "FHS Entropy constant VMPs",
       ylab = "entropy")+
  theme_minimal()



pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_DMPs, Entropy_VMPs, Entropy_all, Entropy_none, Entropy_cVMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('FHS_Entropy all age-related CpGs vs non-age-related CpGs, DMPs, VMPs and constant VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.03662098       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 8.690723e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.156075e-05 

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2459571       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -2.853769e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 2.45909e-05 

#summary statistics constant VMPs
fit <- lm(Entropy_cVMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 8.779314e-17        
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0002427454     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 2.897033e-05


write.table(pheno, "FHS Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE49904 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE49904"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE49904 beta after filtering and imputation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE49904 Phenotypes.txt", sep = '')

ChAMP::champ.SVD(beta = B,
                 pd = pheno %>% select(age,
                                       alcohol_use,
                                       sex,
                                       ethnicity.ch1,
                                       diagnosis,
                                       smoking.history.ch1,
                                       tissue,
                                       weight_lb, sentrix.barcode.ch1),
                 resultsDir = "./CHAMP_SVDimages")

glimpse(pheno)

pheno$ID == colnames(M)

#DMPs
design=model.matrix(~age +
                      sex +
                      ethnicity.ch1 +
                      diagnosis + 
                      smoking.history.ch1,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE49904_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE49904",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE49904_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE49904_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("GSE54730_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE54730 Phenotypes.txt")
rownames(pheno) <- pheno$ID

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#residuals matrix reduced to 23614 CpGs

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg20070090",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('GSE49904_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#105 VMPs with shapiro-wilk

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

tiff('GSE49904_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("Methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE49904.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

### DMPs with CTC 

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE49904 beta after filtering and imputation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE49904 Phenotypes.txt")

#use champ package to obtain cell type estimates 
library(ChAMP)

celltypes <- champ.refbase(beta = B,
                           arraytype = "27K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$ID <- rownames(celltypes)
pheno$ID <- rownames(pheno)

pheno <- left_join(pheno, celltypes, by = "ID")
write.table(pheno, 
            "GSE49904 Phenotypes.txt",
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      sex +
                      ethnicity.ch1 +
                      diagnosis + 
                      smoking.history.ch1 +
                      CD4T +
                      Mono +
                      Bcell + 
                      CD8T +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE49904_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE49904",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE49904_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE49904_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs wtith CTC 

resid <- data.table::fread("GSE49904_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE49904 Phenotypes.txt")
rownames(pheno) <- pheno$ID

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#23469 cpgs in reduced cpg matrix

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg21838233",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE49904_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE49904_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("Methylation at",cpg))
dev.off()


results_BP %>% filter(PVALUE < 0.005)
#76 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE49904.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### Entropy 

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE49904 beta after filtering and imputation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE49904 Phenotypes.txt", sep = " ")

#run limma and adjust for covariets EXCLUDING age

design = model.matrix(~sex +
                        ethnicity.ch1 +
                        diagnosis + 
                        smoking.history.ch1,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$ID <- rownames(Entropy)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy, by = "ID")

#Save entropy plot
tiff('GSE49904_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE49904 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value:0.2402566  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001633674  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001390875 ; 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           ethnicity.ch1 +
                           diagnosis +
                           smoking.history.ch1 +
                           CD4T +
                           CD8T +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$ID <- rownames(Entropy_CTC)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_CTC, by = "ID")

#Save entropy plot
tiff('GSE49904_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE49904 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.8946044 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 1.598239e-05   
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001238985    

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.8774878   

tiff('GSE49904 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8774878",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()
path = "/pvol/Preprocessing/Blood/GEO/GSE49904"
setwd(path)

write.table(pheno, 
            "GSE49904 Phenotypes.txt")


#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$ID <- rownames(Entropy_DMPs)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_DMPs, by = "ID")

tiff('GSE49904_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE49904 Entropy DMPs only")+
  theme_minimal()
dev.off()


#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$ID <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "ID")

tiff('GSE49904_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE49904 Entropy VMPs only")+
  theme_minimal()
dev.off()


pheno_entropy <- pheno %>% select(ID, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Condition",
                                                values_to = "Entropy value")

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
pheno_entropy$Condition <- factor(pheno_entropy$Condition, levels = c("Entropy_DMPs", "Entropy_VMPs", "Entropy"))
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|ID), pheno_entropy)
summary(fit)
# age:ConditionEntropy_VMPs pval 0.0666
# age:ConditionEntropy pval 0.9296            

library(viridis)
tiff('GSE49904_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex +
                        ethnicity.ch1 +
                        diagnosis + 
                        smoking.history.ch1,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$ID <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "ID")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE49904 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE49904 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE49904 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.05102982                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002797124           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001408559       

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.6982522                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 6.44687e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.000165601       


write.table(pheno, "GSE49904 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE80417 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE80417"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE80417 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE80417 Phenotypes raw.txt",sep = "")
rownames(pheno) <- pheno$description
keep = colnames(B)
pheno <- pheno[keep,]
write.table(pheno, "GSE80417 Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

pheno <- read.delim("GSE80417 Phenotypes.txt", sep = "")
glimpse(pheno)
pheno <- pheno %>% select(title, 
                          geo_accession,
                          description,
                          age.ch1,
                          disease.status.ch1,
                          Sex.ch1,
                          tissue.ch1)
names(pheno)[names(pheno) == "age.ch1"] <- "age"
names(pheno)[names(pheno) == "Sex.ch1"] <- "sex"
names(pheno)[names(pheno) == "disease.status.ch1"] <- "disease status"

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$`disease status` <- as.factor(pheno$`disease status`)

library(ChAMP)
champ.SVD(beta = B,
          pd = pheno %>% select(age,
                                sex,
                                `disease status`),
          resultsDir = "./CHAMP_SVDimages")

pheno$Sample_Name == colnames(B)
#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE80417_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#48055 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE80417",
               directory = directory)



#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE80417_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE80417_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("GSE80417_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE80417 Phenotypes.txt", sep = "")
rownames(pheno) <- pheno$description

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#237610 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16722435",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE80417_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#18837 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE80417.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE80417"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE80417_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth*100)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE80417 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE80417 Phenotypes.txt",sep = "\t")

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)

pheno <- pheno %>% select(title, 
                          geo_accession,
                          description,
                          age.ch1,
                          disease.status.ch1,
                          Sex.ch1,
                          tissue.ch1)
names(pheno)[names(pheno) == "age.ch1"] <- "age"
names(pheno)[names(pheno) == "Sex.ch1"] <- "sex"
names(pheno)[names(pheno) == "disease.status.ch1"] <- "disease status"

celltypes$description <- rownames(celltypes)
pheno <- pheno %>% left_join(celltypes, by = "description")

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$`disease status` <- as.factor(pheno$`disease status`)

write.table(pheno, "GSE80417 Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep ="\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE80417.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE80417_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#53383 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE80417",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE80417_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE80417_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC 
resid <- data.table::fread("GSE80417_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE80417 Phenotypes.txt", sep = "\t")
rownames(pheno) <- pheno$description

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#236112 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16722435",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE80417_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#18480 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE80417.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE80417"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE80417_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 
B <- data.table::fread("GSE80417 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE80417 Phenotypes.txt", sep = "")

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- pheno$description
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE80417 Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE80417 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1578088    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 4.108133e-05
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 2.90502e-05 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           CD8T +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC)

#Save entropy plot
tiff('GSE80417_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE80417 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.005250155   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 7.595093e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 2.711586e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.9381497 

tiff('GSE80417 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9381497",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE80417 Phenotypes.txt")

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design=model.matrix(~ sex,
                    pheno)


fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE80417 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE80417 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE80417 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.001604493                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 9.036832e-05            
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 2.851841e-05         

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.4897902                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -2.512914e-05       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.636397e-05        


write.table(pheno, "GSE80417 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### SATSA ####

path = "/pvol/Preprocessing/Blood/ArrayExpress/SATSA"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("SATSA Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("SATSA Phenotypes.txt")

names(pheno)[names(pheno) == "V1"] <- "Sample_Name"
names(pheno)[names(pheno) == "V2"] <- "Extract_Name"
names(pheno)[names(pheno) == "V3"] <- "Organism"
names(pheno)[names(pheno) == "V4"] <- "Individual_ID"
names(pheno)[names(pheno) == "V5"] <- "age"
names(pheno)[names(pheno) == "V6"] <- "unit"
names(pheno)[names(pheno) == "V8"] <- "sex"
names(pheno)[names(pheno) == "V9"] <- "tissue"
names(pheno)[names(pheno) == "V10"] <- "celltype"
names(pheno)[names(pheno) == "V11"] <- "disease.status"
names(pheno)[names(pheno) == "V12"] <- "twin.pair"
names(pheno)[names(pheno) == "V17"] <- "Array.Data.file"
names(pheno)[names(pheno) == "V19"] <- "Array"
names(pheno)[names(pheno) == "V20"] <- "Slide"
names(pheno)[names(pheno) == "V21"] <- "Supplementary"
names(pheno)[names(pheno) == "V22"] <- "Supplementary2"
names(pheno)[names(pheno) == "V25"] <- "predictedSex"
names(pheno)[names(pheno) == "V26"] <- "sexmatch"

write.table(pheno, 
            "SATSA Phenotypes.txt",
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

ChAMP::champ.QC(beta = as.matrix(B),
                pheno = pheno$sex,
                dendrogram = FALSE)

ChAMP::champ.SVD(beta = as.matrix(B),
                 pd = pheno %>% select(age,
                                       sex,
                                       disease.status,
                                       twin.pair,
                                       Slide,
                                       Array),
                 resultsDir = "./SVD_images")

rownames(pheno) <- pheno$Sample_Name

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

#use twin pair as the duplicate correlation
Msubset <- M[sample(nrow(M),size=20000,replace=FALSE),]

corfit <- duplicateCorrelation(Msubset, #we fit limma models on M values and not beta values
                               design,
                               block=pheno$twin.pair) 
corfit$consensus
#0.08156756

fit1_M <- lmFit(M,
                design,
                block = pheno$twin.pair,
                correlation=0.08156756)
fit2_M <- eBayes(fit1_M)

#use twin pair as the duplicate correlation
Bsubset <- B[sample(nrow(M),size=20000,replace=FALSE),]

corfit_B <- duplicateCorrelation(Bsubset, #we fit limma models on M values and not beta values
                                 design,
                                 block=pheno$twin.pair) 
corfit_B$consensus
#0.08154612

fit1_B <- lmFit(B,
                design,
                block = pheno$twin.pair,
                correlation = 0.08154612
)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('SATSA_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "SATSA",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('SATSA_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="SATSA_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("GSE54730_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE54730 Phenotypes.txt")
rownames(pheno) <- pheno$geo_accession

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]


resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg11065154",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

#Save p-value histogram
tiff('SATSA_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#66,664 VMPs without shapiro-wilk

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

tiff('SATSA_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="SATSA.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

### DMPs with CTC 
path = "/pvol/Preprocessing/Blood/ArrayExpress/SATSA"
setwd(path)

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("SATSA Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("SATSA Phenotypes.txt")

#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")
rownames(pheno) <- pheno$Sample_Name

write.table(pheno,
            file="SATSA Phenotypes.txt",
            quote=FALSE,
            row.names=TRUE,
            col.names=TRUE,
            sep='\t')

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_SATSA.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

#use twin pair as the duplicate correlation
Msubset <- M[sample(nrow(M),size=5000,replace=FALSE),]

corfit <- duplicateCorrelation(Msubset, #we fit limma models on M values and not beta values
                               design,
                               block=pheno$twin.pair) 
corfit$consensus
#0.06909469

fit1_M <- lmFit(M,
                design,
                block = pheno$twin.pair,
                correlation=0.06909469)
fit2_M <- eBayes(fit1_M)

#use twin pair as the duplicate correlation
Bsubset <- B[sample(nrow(M),size=5000,replace=FALSE),]

corfit_B <- duplicateCorrelation(Bsubset, #we fit limma models on M values and not beta values
                                 design,
                                 block=pheno$twin.pair) 
corfit_B$consensus
#0.06952708

fit1_B <- lmFit(B,
                design,
                block = pheno$twin.pair,
                correlation = 0.06952708)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('SATSA_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "SATSA",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('SATSA_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="SATSA_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs wtith CTC

resid <- data.table::fread("SATSA_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("SATSA Phenotypes.txt")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]


resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg09846177",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('SATSA_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('SATSA_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()


results_BP %>% filter(PVALUE < 0.005)
#61197 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="SATSA.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### Entropy 

path = "/pvol/Preprocessing/Blood/ArrayExpress/SATSA"
setwd(path)

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("SATSA Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("SATSA Phenotypes.txt", sep = "")

#run limma and adjust for covariates EXCLUDING age

design = model.matrix(~sex,
                      pheno)

#use twin pair as the duplicate correlation
Msubset <- M[sample(nrow(M),size=100,replace=FALSE),]

corfit <- duplicateCorrelation(Msubset, #we fit limma models on M values and not beta values
                               design,
                               block=pheno$twin.pair) 
corfit$consensus
#0.07955328

fit1_M <- lmFit(M,
                design,
                block = pheno$twin.pair,
                correlation=0.07955328)
fit2_M <- eBayes(fit1_M)

resid_M <- residuals(fit2_M, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

which(is.na(B_adj))

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('SATSA_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "SATSA Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()

#needs to be a mixed model with a random intercept
fit <- lmerTest::lmer(Entropy ~ age + (1|twin.pair), pheno)
summary(fit)
pval = summary(fit)$coefficients[,5][2]
pval
#p-value: 0.8347775 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.591866e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 7.628625e-05 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           NK +
                           Bcell + 
                           CD8T +
                           Gran,
                         pheno)

corfit <- duplicateCorrelation(Msubset, #we fit limma models on M values and not beta values
                               designCTC,
                               block=pheno$twin.pair) 
corfit$consensus
#0.07955328

fit1_CTC <- lmFit(M,
                  designCTC,
                  block = pheno$twin.pair,
                  correlation = 0.06572554)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('SATSA_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "SATSA Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lmerTest::lmer(Entropy_CTC ~ age + (1|twin.pair), pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,5][2]
pvalCTC
#p-value: 0.08817062 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001134559 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 6.640959e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.903376 

tiff('SATSA Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.903376",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, 
            "SATSA Phenotypes.txt")


#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_Name <- rownames(Entropy_DMPs)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_DMPs, by = "Sample_Name")

tiff('SATSA_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "SATSA Entropy DMPs only")+
  theme_minimal()
dev.off()


#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$Sample_Name <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "Sample_Name")

tiff('SATSA_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "SATSA Entropy VMPs only")+
  theme_minimal()
dev.off()


pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
pheno_entropy$Condition <- factor(pheno_entropy$Condition, levels = c("Entropy_DMPs", "Entropy_VMPs", "Entropy"))
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|Sample_Name), pheno_entropy)
summary(fit)
# age:ConditionEntropy_VMPs pval 0.0424 *
# age:ConditionEntropy pval 0.8145    

library(viridis)
tiff('SATSA_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)
fit1_M <- lmFit(M_all,
                design,
                block = pheno$twin.pair,
                correlation=0.07955328)
fit2_M <- eBayes(fit1_M)

resid_M <- residuals(fit2_M, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- pheno %>% left_join(Entropy_all, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "SATSA Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design,
              block = pheno$twin.pair,
              correlation = 0.07955328)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)
#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- pheno %>% left_join(Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "SATSA Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('SATSA Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1386175                       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0001125963                
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 7.597254e-05             

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.003085382                         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002438246        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 8.221025e-05         


write.table(pheno, "SATSA Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




#### GSE42861 ####
path = "/pvol/Preprocessing/Blood/GEO/GSE42861"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE42861 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE42861 Phenotypes.txt", sep = "\t")

glimpse(pheno)

rownames(pheno) <- pheno$V1
names(pheno)[names(pheno)=="V1"] <- "Sample_Name"
names(pheno)[names(pheno)=="V2"] <- "title"
names(pheno)[names(pheno)=="V3"] <- "sample info"
names(pheno)[names(pheno)=="V4"] <- "geo_accession"
names(pheno)[names(pheno)=="V5"] <- "supplementary_file"
names(pheno)[names(pheno)=="V6"] <- "IDAT colour"
names(pheno)[names(pheno)=="V7"] <- "age"
names(pheno)[names(pheno)=="V8"] <- "cell type"
names(pheno)[names(pheno)=="V9"] <- "disease state"
names(pheno)[names(pheno)=="V10"] <- "sex"
names(pheno)[names(pheno)=="V11"] <- "smoking status"
names(pheno)[names(pheno)=="V12"] <- "subject"
names(pheno)[names(pheno)=="V13"] <- "Array"
names(pheno)[names(pheno)=="V14"] <- "Slide"
names(pheno)[names(pheno)=="V15"] <- "Basename"
names(pheno)[names(pheno)=="V16"] <- "Basename2"
names(pheno)[names(pheno)=="V19"] <- "predictedSex"
names(pheno)[names(pheno)=="V20"] <- "sexmatch"

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$disease.state <- as.factor(pheno$disease.state)


#DMPs
design=model.matrix(~age +
                      sex +
                      disease.state,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE42861_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#65328 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE42861",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE42861_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE42861_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("GSE42861_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE42861 Phenotypes.txt", sep = "\t")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#302289 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg25104511",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE42861_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#36783 VMPs  

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE42861.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE42861"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE42861_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE42861 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE42861 Phenotypes.txt", sep = "\t")

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- pheno %>% left_join(celltypes, by = "Sample_Name")

write.table(pheno, "GSE42861 Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep ="\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex +
                                            disease.state,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE42861.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      disease.state +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE42861_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#50863 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE42861",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE42861_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE42861_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC 
resid <- data.table::fread("GSE42861_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE42861 Phenotypes.txt", sep = "\t")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#296707 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16722435",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE42861_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#23712 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE42861.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE42861"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE42861_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy
B <- data.table::fread("GSE42861 Normalized beta values after batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE42861 Phenotypes.txt", sep = "")

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex +
                        disease.state,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE42861 Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE42861 Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.38,0.51)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.001366545   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001592337
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.953171e-05  

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           disease.state +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC)

#Save entropy plot
tiff('GSE42861_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE42861 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.39,0.49)
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.04421635   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 8.123439e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 4.03001e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.8086931 

tiff('GSE42861 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8086931  ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE42861 Phenotypes.txt")

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex + disease.state,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE42861 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE42861 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE42861 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 2.063204e-06                
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003073444           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.418927e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.3254957                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -4.588918e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.663861e-05      


write.table(pheno, "GSE42861 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE51032 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE51032"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE51032 filtered and imputed.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE51032 Phenotypes normal samples only.txt")

glimpse(pheno)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE51032_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#21863 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE51032",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE51032_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE51032_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs
resid <- data.table::fread("GSE51032_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE51032 Phenotypes normal samples only.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]

resid_B <- minfi::ilogit2(resid)
#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]
fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg11233000",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE51032_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3883 VMPs with adj pval < 0.005

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE51032.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE51032"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE51032_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE51032 filtered and imputed.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% dplyr::select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE51032 Phenotypes normal samples only.txt")

#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$geo_accession <- rownames(celltypes)
pheno$geo_accession <- as.character(pheno$geo_accession)
pheno <- left_join(pheno, celltypes, by = "geo_accession")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE51032.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      CD4T +
                      Bcell +
                      CD8T +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE51032_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#22108 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE51032",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE51032_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE51032_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")
write.table(pheno,
            file="GSE51032 Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- data.table::fread("GSE51032_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE51032 Phenotypes for normal samples only.txt")
rownames(pheno) <- pheno$geo_accession

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE51032_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#4085 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE51032.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE51032"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE51032_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy 

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE51032 filtered and imputed.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% dplyr::select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE51032 Phenotypes.txt")

B[B==0] <- min(B[B>0])

#run limma and adjust for covariates EXCLUDING age
M <- logit2(B)

design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$geo_accession <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "geo_accession")

#Save entropy plot
tiff('GSE51032_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE51032 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.001526438  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0004515766 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001414877 

#Entropy with CTC 

designCTC = model.matrix(~ sex +
                           CD4T +
                           NK +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$geo_accession<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "geo_accession")

#Save entropy plot
tiff('GSE5103_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE51032 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.001150322   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0004008535  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001224137   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7915324 

tiff('GSE51032 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.791",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE51032 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### DMPs vs VMPs entropy 

#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$title <- rownames(Entropy_DMPs)
pheno <- left_join(pheno, Entropy_DMPs, by = "title")

tiff('GSE51032_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE51032 Entropy DMPs only")+
  theme_minimal()
dev.off()

fit <- lm(Entropy_DMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#8.41787e-21 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 9.500367e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 1.007322e-05

#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$title <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "title")

tiff('GSE51032_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE51032 Entropy VMPs only")+
  theme_minimal()
dev.off()

fit2 <- lm(Entropy_VMPs ~ age, pheno)
summary(fit2)
pval2 = summary(fit2)$coefficients[,4][2]
pval2
#5.220734e-36 
effect2 = summary(fit2)$coefficients[,1][2]
effect2
#effect: 0.0003081715 
stderr2 = summary(fit2)$coefficients[,2][2]
stderr2
#stderr = 2.423959e-05

pheno_entropy <- pheno %>% select(title, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,title),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|title), pheno_entropy)
summary(fit)
# age:ConditionEntropy_VMPs pval < 2e-16 ***
# ge:ConditionEntropy_DMPs pval 0.368

tiff('GSE51032_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Class, alpha = 0.8)) +
  geom_smooth(aes(colour = Class), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()





#### Correlation of age and cell types

cor.test(pheno$age, pheno$CD8T, method = "pearson")
#-0.05

ggplot(pheno) +
  geom_jitter(aes(x = age, y = CD8T), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = CD8T), colour = "black") +
  labs(title = "Pearson's corr = -0.05")+
  theme_minimal()

cor.test(pheno$age, pheno$CD4T, method = "pearson")
#0.07

ggplot(pheno) +
  geom_jitter(aes(x = age, y = CD4T), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = CD4T), colour = "black") +
  labs(title = "Pearson's corr = 0.07")+
  theme_minimal()


cor.test(pheno$age, pheno$Bcell, method = "pearson")
#0.05

ggplot(pheno) +
  geom_jitter(aes(x = age, y = Bcell), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = Bcell), colour = "black") +
  labs(title = "Pearson's corr = 0.05")+
  theme_minimal()


cor.test(pheno$age, pheno$NK, method = "pearson")
#0.1

ggplot(pheno) +
  geom_jitter(aes(x = age, y = NK), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = NK), colour = "black") +
  labs(title = "Pearson's corr = 0.1")+
  theme_minimal()

cor.test(pheno$age, pheno$Mono, method = "pearson")
#-0.01

ggplot(pheno) +
  geom_jitter(aes(x = age, y = Mono), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = Mono), colour = "black") +
  labs(title = "Pearson's corr = -0.01")+
  theme_minimal()

cor.test(pheno$age, pheno$Gran, method = "pearson")
#-0.089

ggplot(pheno) +
  geom_jitter(aes(x = age, y = Gran), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = Gran), colour = "black") +
  labs(title = "Pearson's corr = -0.089")+
  theme_minimal()


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$geo_accession <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "geo_accession")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE51032 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE51032 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE51032 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.000100832                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0006004967           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001528715       

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1682637                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002151578         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.000155877        


write.table(pheno, "GSE51032 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE67705 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE67705"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE67705 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE67705 Phenotypes.txt")


pheno <- separate(pheno, col = HIV.status, into = c("HIV.status","Patient", "patient ID"), sep = "\\ ")
pheno$HIV.status <- as.factor(pheno$HIV.status)

#DMPs
design=model.matrix(~age +
                      HIV.status,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)

fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE67705_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE67705",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE67705_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="
            _M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


pheno %>% summarise(mean = mean(age),
                    sd = sd(age),
                    range= range(age))
#### VMPs 
resid <- data.table::fread("GSE67705_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE67705 Phenotypes.tst")
rownames(pheno) <- pheno$geo_accession

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#reduced residuals matrix to 294046 cpgs

resid_B <- ilogit2(resid)
#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg20629239",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE67705_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2536 VMPs

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE67705.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE67705"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE67705_VMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()


#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE67705 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE67705 Phenotypes.txt")

pheno <- separate(pheno, col = HIV.status, into = c("HIV.status","Patient", "patient ID"), sep = "\\ ")
pheno$HIV.status <- as.factor(pheno$HIV.status)

write.table(pheno,
            "GSE67705 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE,
            row.names = TRUE)

#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$geo_accession <- rownames(celltypes)
pheno$geo_accession <- as.character(pheno$geo_accession)
pheno <- left_join(pheno, celltypes, by = "geo_accession")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~HIV.status,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE67705.rds')

#DMPs
design=model.matrix(~age +
                      HIV.status +
                      CD4T +
                      Bcell +
                      Mono +
                      CD8T +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE67705_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#10524 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE67705",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE67705_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE67705_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")
write.table(pheno,
            file="GSE67705 Phenotypes.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC 
resid <- data.table::fread("GSE67705_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE67705 Phenotypes.txt")
rownames(pheno) <- pheno$geo_accession

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE67705_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2638 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE67705.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE67705"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE67705_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()


#### Entropy 

#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE67705 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE67705 Phenotypes.txt")

#run limma and adjust for covariates EXCLUDING age
design = model.matrix(~HIV.status,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

which(is.na(B_adj))

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$geo_accession <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "geo_accession")

#Save entropy plot
tiff('GSE67705_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67705 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.00380295   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 8.385853e-05  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 2.860733e-05  

#Entropy with CTC 

designCTC = model.matrix(~HIV.status +
                           CD4T +
                           NK +
                           Bcell + 
                           Mono +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

which(is.na(B_CTC_adj))

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$geo_accession<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "geo_accession")

#Save entropy plot
tiff('GSE67705_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67705 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.001462808 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001567062  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 2.611274e-05   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8893037    

tiff('GSE67705 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8893037",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE67705 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### DMPs vs VMPs entropy 

#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$geo_accession <- rownames(Entropy_DMPs)
pheno <- left_join(pheno, Entropy_DMPs, by = "geo_accession")

tiff('GSE67705_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67705 Entropy DMPs only")+
  theme_minimal()
dev.off()


#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$geo_accession <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "geo_accession")

tiff('GSE67705_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67705 Entropy VMPs only")+
  theme_minimal()
dev.off()


pheno_entropy <- pheno %>% select(geo_accession, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,geo_accession),
                                                names_to = "Condition",
                                                values_to = "Entropy value")

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
pheno_entropy <- pheno_entropy %>% mutate(Condirion = factor(Condition, levels = c("Entropy_DMPs","Entropy_VMPs","Entropy")))
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|geo_accession), pheno_entropy)
summary(fit)
# age:ConditionEntropy_VMPs pval 0.011796 *  
# ge:ConditionEntropy_DMPs pval 0.895991
library(viridis)
tiff('GSE67705_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~HIV.status,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$geo_accession <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "geo_accession")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67705 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67705 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE67705 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.0001655302                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001432541          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.72504e-05     

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.9221143                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 4.148647e-06        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.23742e-05      


write.table(pheno, "GSE67705 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




#### GSE32148 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE32148"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE32148 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE32148 pheno.txt")

library(ChAMP)
B <- as.matrix(B)
champ.QC(beta = B, 
         pheno = pheno$sex,
         dendrogram = FALSE)
champ.SVD(beta = B,
          pd = pheno %>% select(age,
                                disease.state,
                                sex,
                                twin),
          resultsDir = "./CHAMP_SVDimages")

#DMPs
design=model.matrix(~age +
                      sex +
                      disease.state,
                    pheno)

#corfit <- duplicateCorrelation(M, #we fit limma models on M values and not beta values
#                               design,
#                               block=pheno$twin) 
#corfit$consensus
#0

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE32148_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#2667 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE32148",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE32148_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE32148_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs
resid <- read.table("GSE32148_M_res.txt")
pheno <- read.delim("GSE32148 pheno.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#365605 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg10320884",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE32148_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2643 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE32148.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE32148"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE32148_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth, colour = twin)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE32148 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE32148 pheno.txt")


#DMPs
design=model.matrix(~age +
                      sex +
                      disease.state + 
                      CD4T +
                      Bcell +
                      Mono +
                      CD8T +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE32148_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#3606 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE32148",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE32148_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE32148_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE32148_M_CTC_res.txt")
pheno <- read.delim("GSE32148 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#367352 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE32148_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#5291 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE32148.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE32148"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE32148_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE32148 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE32148 pheno.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        disease.state,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$geo_accession <- rownames(Entropy)
pheno$geo_accession <- rownames(pheno)
pheno <- left_join(pheno, Entropy, by = "geo_accession")

#Save entropy plot
tiff('GSE32148_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE32148 Entropy",
       ylab = "entropy")+
  theme_minimal() +
  ylim(0.35,0.5)
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.03031672
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 7.717136e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 3.553478e-05 

#Entropy with CTC 

designCTC = model.matrix(~CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$title<- rownames(Entropy_CTC)
pheno$title
pheno <- left_join(pheno, Entropy_CTC, by = "title")

#Save entropy plot
tiff('GSE32148_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE32148 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.002983527  
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 9.377501e-05 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 3.143504e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8911446 

tiff('GSE32148 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.891",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE32148 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex + disease.state,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

B_adj[B_adj == 0] <- min(B_adj[which(B_adj > 0)])
B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$geo_accession <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "geo_accession")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE32148 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

B_adj[B_adj == 0] <- min(B_adj[which(B_adj > 0)])
B_adj <- na.omit(B_adj)


#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$geo_accession <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "geo_accession")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE32148 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE32148 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.6639592                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 6.735856e-05          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001539472       

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5142416                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 7.290612e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001108295       


write.table(pheno, "GSE32148 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE106648 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE106648"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE106648 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE106648 Phenotypes.txt")
glimpse(pheno)


#DMPs
design=model.matrix(~age +
                      sex +
                      smoking.status,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE106648_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#29570 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE106648",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE106648_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE106648_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE106648_M_res.txt")
pheno <- read.delim("GSE106648 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#353581 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE106648_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#10447 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE106648.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE106648"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE106648_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE106648 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE106648 Phenotypes.txt")
glimpse(pheno)

names(pheno)[names(pheno)=="title"] <- "Sample_Name"
rownames(pheno) <- pheno$Sample_Name

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE106648 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            smoking.status,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE106648.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      smoking.status +
                      CD4T +
                      CD8T +
                      Mono +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE106648_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#53496 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE106648",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE106648_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE106648_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE106648_M_CTC_res.txt")
pheno <- read.delim("GSE106648 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE106648_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#8557 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE106648.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE106648"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE106648_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE106648 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE106648 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        smoking.status,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE106648_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE106648 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.0005464461   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001655392    
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.73277e-05 


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           smoking.status +
                           CD4T +
                           CD8T +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE106648_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE106648 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 5.869897e-08   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001908807 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 3.423793e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7494356   

tiff('GSE106648 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7494356",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE106648 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)



#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        smoking.status,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE106648 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE106648 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE106648 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.0001197673           
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002837366       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 7.271297e-05  

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8445477            
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 6.917006e-06        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.524344e-05   


write.table(pheno, "GSE106648 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE69138 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE69138"
setwd(path)

library(minfi)
library(tidyverse)
library(limma)
library(data.table)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- fread("GSE69138 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE69138 Phenotypes.txt")
glimpse(pheno)

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$stoke.subtype <- as.factor(pheno$stroke.subtype)

#DMPs
design=model.matrix(~age +
                      sex + 
                      stroke.subtype,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE69138_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#975 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE69138",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE69138_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE69138_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs
resid <- read.table("GSE69138_M_res.txt")
pheno <- read.delim("GS69138 Phenotypes.txt", sep = "\t")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#390145 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg15742840",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE69138_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#8732 VMPs  

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE69138.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE69138"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE69138_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg)) +
  ylim(0.5,1)
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data
library(data.table)
B <- fread("GSE69138 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE69138 Phenotypes.txt")
glimpse(pheno)

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$stoke.subtype <- as.factor(pheno$stroke.subtype)

library(ChAMP)

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)

pheno <- left_join(pheno, celltypes, by = "Sample_Name")

glimpse(pheno)
write.table(pheno, "GSE69138 Phenotypes.txt",
            quote = FALSE,
            col.names = TRUE,
            sep = "\t")

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)
pheno$CD4T <- as.numeric(pheno$CD4T)
pheno$CD8T <- as.numeric(pheno$CD8T)
pheno$NK <- as.numeric(pheno$NK)
pheno$Gran <- as.numeric(pheno$Gran)
pheno$Mono <- as.numeric(pheno$Mono)
pheno$Bcell <- as.numeric(pheno$Bcell)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex +
                                            stroke.subtype,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE69138.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      stroke.subtype +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
#   CD4T +
#   Bcell +
#   Mono +
#   NK +
#   Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE69138_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#129 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"
#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE69138",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE69138_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE69138_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE69138_M_CTC_res.txt")
pheno <- read.delim("GSE69138 Phenotypes.txt", sep = "\t")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#389325 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg09791440",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE69138_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#7598 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE69138.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE69138"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE69138_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

ages_range <- ggplot(data = pheno) +
  geom_histogram(mapping = aes(x = age), colour = "black", fill = "skyblue", bins = 40) + 
  theme_minimal()
ggsave("age distribution.tiff")

#### Entropy
path = "/pvol/Preprocessing/Blood/GEO/GSE69138"
setwd(path)
#load the residuals from the DMP model adjusted for age and covariates 
B <- data.table::fread("GSE69138 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE69138 Phenotypes.txt", sep = "")

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex +
                        stroke.subtype,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE69138_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black", se = FALSE) +
  labs(title = "GSE69138 Entropy",
       ylab = "entropy")+
  theme_minimal()+
  ylim(0.38, 0.5)
dev.off()

fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8784567
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -1.707746e-05  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr =0.0001115146  

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           stroke.subtype +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE69138_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE69138 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()+
  ylim(0.38, 0.5)
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.9707839
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 3.056264e-06  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 8.333243e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.7484675 

tiff('GSE69138 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7484675",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE69138 Phenotypes.txt")



#### DMPs vs VMPs entropy 

#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$Sample_Name <- rownames(Entropy_DMPs)
pheno <- left_join(pheno, Entropy_DMPs, by = "Sample_Name")

tiff('GSE55763_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE55763 Entropy DMPs only")+
  theme_minimal()
dev.off()

fit <- lm(Entropy_DMPs ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.725253 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 4.258091e-05  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001209719

#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$Sample_Name <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "Sample_Name")

tiff('GSE55763_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE55763 Entropy VMPs only")+
  theme_minimal()
dev.off()

fit2 <- lm(Entropy_VMPs ~ age, pheno)
summary(fit2)
pval2 = summary(fit2)$coefficients[,4][2]
pval2
#0.4786825  
effect2 = summary(fit2)$coefficients[,1][2]
effect2
#effect:-0.0001289632
stderr2 = summary(fit2)$coefficients[,2][2]
stderr2
#stderr = 0.0001816672

pheno_entropy <- pheno %>% select(age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!age,
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Class"

library(viridis)
tiff('GSE55763_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Class, alpha = 0.8)) +
  geom_smooth(aes(colour = Class), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()

write.table(pheno,
            "GSE69138 Phenotypes.txt")


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex + stoke.subtype,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE69138 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE69138 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE69138 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.6841251              
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -6.065552e-05          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001488506     

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.7793944               
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 3.058657e-05       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001090338     


write.table(pheno, "GSE69138 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE50660 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE50660"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE50660 beta filtered.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE50660 Phenotypes.txt")
glimpse(pheno)

names(pheno)[names(pheno) == "geo_accession"] <- "Sample_Name"

#DMPs
design=model.matrix(~age +
                      sex +
                      `smoking..0..1.and.2..which.represent.never..former.and.current.smokers.`,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE50660_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#2654 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE50660",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE50660_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE50660_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE50660_M_res.txt")
pheno <- read.delim("GSE50660 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#211084 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg01188146",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE50660_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2645 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE50660.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE50660"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE50660_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg)) 
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE50660 beta filtered.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE50660 Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE50660 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex +
                                            `smoking..0..1.and.2..which.represent.never..former.and.current.smokers.`,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE50660.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      `smoking..0..1.and.2..which.represent.never..former.and.current.smokers.` +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE50660_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#11631 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE50660",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE50660_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE50660_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- read.table("GSE50660_M_CTC_res.txt")
pheno <- read.delim("GSE50660 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE50660_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2762 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE50660.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE50660"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE50660_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE50660 beta filtered.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE50660 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        `smoking..0..1.and.2..which.represent.never..former.and.current.smokers.`,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE50660_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE50660 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.6890406   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -4.478017e-05   
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.000111836 


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           `smoking..0..1.and.2..which.represent.never..former.and.current.smokers.` +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE50660_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE50660 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.9971298   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 3.689511e-07 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001025072  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.9176421   

tiff('GSE50660 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9176421 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE50660 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex +`smoking..0..1.and.2..which.represent.never..former.and.current.smokers.`,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE50660 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE50660 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE50660 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.947926                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 7.245851e-06          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001108826      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.3793442                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.000114795        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001304561       


write.table(pheno, "GSE50660 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




#### GSE40279 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE40279"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE40279 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE40279 Phenotypes updated.txt", sep = "")
rownames(pheno) <- pheno$ID

#DMPs
design=model.matrix(~age +
                      sex +
                      ethnicity,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE40279_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE40279",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE40279_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE40279_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs
resid <- data.table::fread("GSE40279_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE40279 Phenotypes updated.txt")
rownames(pheno) <- pheno$ID

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
#residuals matrix reduced to 148915 CpGs

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg02734358",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

results_metaRE <- tibble(CPG = rownames(results),
                         TESTSTAT = results$TestStat,
                         N = rep(ncol(w),nrow(w)),
                         COEF = results$logFC,
                         SE = results$SE,
                         PVALUE = results$chisq_pval,
                         BP = ncol(resid)*fitted_sqrd/rowSums(w^2))

#Save p-value histogram
tiff('GSE40279_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#76774 VMPs with shapiro-wilk

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(resid_B[cpg,]))

tiff('GSE40279_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=(meth^2)*100)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE40279.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

setwd("/pvol/EWAS/Blood/RE meta")
write.table(results_metaRE,
            file="GSE40279_VMPs.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### DMPs CTC
path = "/pvol/Preprocessing/Blood/GEO/GSE40279"
setwd(path)

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GSE40279 beta after filtering and imputation.txt")
M <- logit2(B)
pheno <- read.delim("GSE40279 Phenotypes updated.txt", sep = "")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
ID <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(ID = ID)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex+
                                            ethnicity,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE40279.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      sex +
                      ethnicity +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE40279_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"

create_summary(toptable = results,
               dataset_label = "GSE40279",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE40279_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE40279_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC
resid <- data.table::fread("GSE40279_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE40279 Phenotypes updated.txt")
rownames(pheno) <- pheno$ID

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
z#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
#175443 cpgs in reduced cpg matrix

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg13066289",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE40279_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
#effect <- results_BP$COEF[1]
tiff('GSE40279_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()

results_BP %>% filter(PVALUE < 0.005)
#70187 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE40279.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### Entropy 
B <- data.table::fread("GSE40279 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- select(B, -V1)
pheno <- read.delim("GSE40279 Phenotypes updated.txt", sep = "")

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex+
                        ethnicity,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)


#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$ID <- rownames(Entropy)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy, by = "ID")

#Save entropy plot
tiff('GSE40279_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.3204021
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 5.105538e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 5.134355e-05

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           ethnicity +
                           CD4T +
                           NK +
                           B + 
                           Neutro+
                           Mono,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$ID <- rownames(Entropy_CTC)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_CTC, by = "ID")

#Save entropy plot
tiff('GSE40279_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.9272555 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 4.245417e-06
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 4.648252e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.9235871 

tiff('GSE40279 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9235871",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, 
            "GSE40279 Phenotypes updated.txt")


#### DMPs vs VMPs entropy 

#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$ID <- rownames(Entropy_DMPs)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_DMPs, by = "ID")

tiff('GSE40279_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy DMPs only")+
  theme_minimal()
dev.off()

#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$ID <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "ID")

tiff('GSE40279_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy VMPs only")+
  theme_minimal()
dev.off()

pheno_entropy <- pheno %>% select(ID, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|ID), pheno_entropy)
summary(fit)

library(viridis)
tiff('GSE40279_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()





#### Entropy all vs non

#load the data 
B <- data.table::fread("GSE40279 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE40279 Phenotypes updated.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex+
                        ethnicity,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$ID <- rownames(Entropy_all)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_all, by = "ID")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        ethnicity,
                      pheno)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$ID <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "ID")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


pheno_entropy <- pheno %>% select(ID, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE40279 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.09228069         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 9.913638e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.880118e-05 

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5737251          
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -3.079997e-05      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.472078e-05  


write.table(pheno, "GSE40279 Phenotypes updated.txt",
            sep = "\t",
            col.names = TRUE)
path = "/pvol/Preprocessing/Blood/GEO/GSE40279"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE40279 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE40279 Phenotypes updated.txt", sep = "")
rownames(pheno) <- pheno$ID

#DMPs
design=model.matrix(~age +
                      sex +
                      ethnicity,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE40279_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE40279",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE40279_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE40279_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- data.table::fread("GSE40279_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE40279 Phenotypes updated.txt")
rownames(pheno) <- pheno$ID

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
#residuals matrix reduced to 148915 CpGs

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg02734358",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

results_metaRE <- tibble(CPG = rownames(results),
                         TESTSTAT = results$TestStat,
                         N = rep(ncol(w),nrow(w)),
                         COEF = results$logFC,
                         SE = results$SE,
                         PVALUE = results$chisq_pval,
                         BP = ncol(resid)*fitted_sqrd/rowSums(w^2))

#Save p-value histogram
tiff('GSE40279_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#76774 VMPs with shapiro-wilk

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(resid_B[cpg,]))

tiff('GSE40279_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)

ggplot(pheno_with_meth, aes(x=age, y=(meth^2)*100)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE40279.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

setwd("/pvol/EWAS/Blood/RE meta")
write.table(results_metaRE,
            file="GSE40279_VMPs.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### DMPs CTC
path = "/pvol/Preprocessing/Blood/GEO/GSE40279"
setwd(path)

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GSE40279 beta after filtering and imputation.txt")
M <- logit2(B)
pheno <- read.delim("GSE40279 Phenotypes updated.txt", sep = "")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
ID <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(ID = ID)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex+
                                            ethnicity,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE40279.rds')

#DMPs with cell type correction (CTC)
design=model.matrix(~age +
                      sex +
                      ethnicity +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
                    # CD4T +
                    # Bcell +
                    # CD8T +
                    # NK +
                    # Gran,
                    # pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE40279_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"

create_summary(toptable = results,
               dataset_label = "GSE40279",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE40279_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE40279_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC
resid <- data.table::fread("GSE40279_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE40279 Phenotypes updated.txt")
rownames(pheno) <- pheno$ID

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
z#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
#175443 cpgs in reduced cpg matrix

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg13066289",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE40279_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
#effect <- results_BP$COEF[1]
tiff('GSE40279_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("% methylation at",cpg))
dev.off()

results_BP %>% filter(PVALUE < 0.005)
#70187 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE40279.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

#### Entropy 
B <- data.table::fread("GSE40279 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- select(B, -V1)
pheno <- read.delim("GSE40279 Phenotypes updated.txt", sep = "")

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex+
                        ethnicity,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)


#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$ID <- rownames(Entropy)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy, by = "ID")

#Save entropy plot
tiff('GSE40279_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.3204021
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 5.105538e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 5.134355e-05

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           ethnicity +
                           CD4T +
                           NK +
                           B + 
                           Neutro+
                           Mono,
                         pheno)

fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$ID <- rownames(Entropy_CTC)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_CTC, by = "ID")

#Save entropy plot
tiff('GSE40279_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()

fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.9272555 
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 4.245417e-06
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 4.648252e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  2.2e-16
#cor = 0.9235871 

tiff('GSE40279 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9235871",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, 
            "GSE40279 Phenotypes updated.txt")


#### DMPs vs VMPs entropy 

#DMPs

DMPs <- read.delim("/pvol/Preprocessing/Blood/Meta_DMPs.txt", sep = "")
DMPs <- DMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_DMPs <- B_adj[rownames(B_adj) %in% DMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_DMPs <- apply(B_adj_DMPs, 2, calculate_entropy)
Entropy_DMPs <- as.data.frame(Entropy_DMPs)
Entropy_DMPs$ID <- rownames(Entropy_DMPs)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_DMPs, by = "ID")

tiff('GSE40279_Entropy_DMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_DMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy DMPs only")+
  theme_minimal()
dev.off()

#VMPs

VMPs <- read.delim("/pvol/Preprocessing/Blood/Meta VMPs.txt", sep = "")
VMPs <- VMPs$MarkerName
B_adj <- as.data.frame(B_adj) 
B_adj_VMPs <- B_adj[rownames(B_adj) %in% VMPs,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_VMPs <- apply(B_adj_VMPs, 2, calculate_entropy)
Entropy_VMPs <- as.data.frame(Entropy_VMPs)
Entropy_VMPs$ID <- rownames(Entropy_VMPs)
pheno <- left_join(pheno, Entropy_VMPs, by = "ID")

tiff('GSE40279_Entropy_VMPsonly.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_VMPs)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy VMPs only")+
  theme_minimal()
dev.off()

pheno_entropy <- pheno %>% select(ID, age, Entropy, Entropy_DMPs, Entropy_VMPs)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

#Entropy ~ Age + Condition + Age x Condition + (1|ID)
library(lmerTest)
fit <- lmerTest::lmer(`Entropy value` ~ age + Condition + age*Condition + (1|ID), pheno_entropy)
summary(fit)

library(viridis)
tiff('GSE40279_Entropy all, DMPs and VMPs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
pheno_entropy <- pheno_entropy %>% mutate(Condition = factor(Condition, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()





#### Entropy all vs non 

#load the data 
B <- data.table::fread("GSE40279 beta after filtering and imputation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE40279 Phenotypes updated.txt", sep = "")

DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for covariets EXCLUDING age
design = model.matrix(~sex+
                        ethnicity,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$ID <- rownames(Entropy_all)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_all, by = "ID")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        ethnicity,
                      pheno)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$ID <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "ID")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE40279 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()


pheno_entropy <- pheno %>% select(ID, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,ID),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE40279 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.09228069         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 9.913638e-05     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.880118e-05 

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5737251          
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -3.079997e-05      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.472078e-05  


write.table(pheno, "GSE40279 Phenotypes updated.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE41037 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE41037"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)

#### DMPs
#Create function to format results for METAL software
create_summary <- function(toptable = NULL,
                           dataset_label= NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#Load data
B <- as.matrix(read.table("GSE41037_filtered_batch_corrected_beta.txt"))
library(minfi)
M <- logit2(B)

library(tidyverse)
pheno <- read_delim("GSE41037_phenotypes.txt",
                    lazy=TRUE)

nrow(pheno)
mean(pheno$age)
sd(pheno$age)
range(pheno$age)
sum(pheno$gender=="male")/nrow(pheno)

#EWAS
design=model.matrix(~age+
                      gender,
                    pheno)
library(limma)
fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)
fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE41037_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE41037",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE41037_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

cpg <- as.character(rownames(results)[1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE41037_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE41037_M_res.txt")
pheno <- read.delim("GSE41037 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE41037_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#1030 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE41037.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE41037"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE41037_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE41037_filtered_batch_corrected_beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE41037 Phenotypes.txt")
glimpse(pheno)

library(ChAMP)
celltypes <- champ.refbase(beta = B)
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno$Sample_Name <- pheno$`GEO accession`
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE41037 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#DMPs
design=model.matrix(~age +
                      gender +
                      CD4T +
                      Bcell +
                      Mono +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE41037_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#2245 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE41037",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE41037_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE41037_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC
resid <- read.table("GSE41037_M_CTC_res.txt")
pheno <- read.delim("GSE41037 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#311219 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE41037_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#838 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE41037.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE41037"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE41037_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE41037_filtered_batch_corrected_beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE41037 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~gender,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE41037_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE41037 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.1592406      
effect = summary(fit)$coefficients[,1][2]
effect
#effect:4.558367e-05
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr =3.234884e-05 


#Entropy with CTC 

designCTC = model.matrix(~gender +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE41037_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE41037 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.008152934       
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 6.725139e-05 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 2.534691e-05   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7908373      

tiff('GSE41037 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7908373",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE41037 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~gender,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE41037 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE41037 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE41037 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.0005914153                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001207731           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.499379e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.5985051                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -1.809172e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.434323e-05      


write.table(pheno, "GSE41037 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE41169 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE41169"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE41169 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE41169 Phenotypes.txt")
glimpse(pheno)

names(pheno)[names(pheno) == "geo_accession"] <- "Sample_Name"

pheno$Sample_Name == colnames(M)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE41169_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#4453 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE41169",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE41169_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE41169_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE41169_M_res.txt")
pheno <- read.delim("GSE41169 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#354276 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg11015188",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE41169_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#4,5119 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE41169.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE41169"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE41169_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg)) 
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE41169 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE41169 Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE41169 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE41169.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      CD4T +
                      Bcell +
                      Mono +
                      CD8T +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE41169_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#16767 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE41169",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE41169_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE41169_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- read.table("GSE41169_M_CTC_res.txt")
pheno <- read.delim("GSE41169 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- as.data.frame(resid)
resid <- resid[CpGs_to_keep,]
#356046 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE41169_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#4000 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE41169.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE41169"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE41169_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE41169 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE41169 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

which(is.infinite(resid_M))
row.is.infinite <- apply(resid_M, 1, function(x){any(is.infinite(x))})
resid_M <- resid_M[!row.is.infinite,]

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M[rownames(resid_M),])
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE41169_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE41169 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.2869219    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001347177    
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001257659  


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)


which(is.infinite(resid_M_CTC))
row.is.infinite <- apply(resid_M_CTC, 1, function(x){any(is.infinite(x))})
resid_M_CTC <- resid_M_CTC[!row.is.infinite,]

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_CTC_adj <- resid_M_CTC + rowMeans(M[rownames(resid_M_CTC),])
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE41169_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE41169 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.06228045    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002208513 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001170074  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.9453665    

tiff('GSE41169 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9453665",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE41169 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

B_adj[B_adj ==0] <- min(B_adj[which(B_adj > 0)])

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}


Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE41169 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197676 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

B_adj[B_adj ==0] <- min(B_adj[which(B_adj >0)])
#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197676 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE197676 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1035785                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002336071           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001420766       

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.9900006                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.854845e-06        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001475941       


write.table(pheno, "GSE41169 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE53840 ####

#Find DMPs
path = "/pvol/Preprocessing/Blood/GEO/GSE53840"
setwd(path)

library(limma)
library(tidyverse)
library(minfi)

#### DMPs 
#Create function to format results for METAL software
create_summary <- function(toptable = NULL,
                           dataset_label= NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#Load data
B <- read.table("GSE53840_normalized_batch_corrected_beta.txt")
library(minfi)
M <- logit2(B)
library(tidyverse)
pheno <- read.delim("GSE53840 Phenotypes.txt")
nrow(pheno)

library(limma)

names(pheno)[names(pheno) == "age (yrs)"] <- "age"
glimpse(pheno)

#EWAS
design=model.matrix(~age+
                      hiv.viral.load,
                    pheno)

pheno$Sample_Name <- pheno$`GEO accession`
pheno$Sample_Name == colnames(B)
pheno <- pheno %>% filter(!is.na(hiv.viral.load))
B <- B[,pheno$Sample_Name]
M <- M[,pheno$Sample_Name]

write.table(pheno,"GSE53840 Phenotypes.txt")
write.table(B, 
            "GSE53840_normalized_batch_corrected_beta.txt",
            row.names = TRUE,
            col.names = TRUE,
            sep = "\t")

library(limma)
fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)
fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE53840_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE53840",
               directory = directory)
#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE53840_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs 
resid <- read.table("GSE53840_M_res.txt")
pheno <- read.delim("GSE53840 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE53840_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#5147 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE53840.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE53840"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE53840_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()



#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- read.table("GSE53840_normalized_batch_corrected_beta.txt")
library(minfi)
M <- logit2(B)
library(tidyverse)
pheno <- read.delim("GSE53840 Phenotypes.txt")

library(ChAMP)
celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE53840 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~hiv.viral.load,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE53840.rds')

#DMPs
design=model.matrix(~age +
                      hiv.viral.load+ 
                      CD4T +
                      Bcell +
                      Mono +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE53840_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#21 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE53840",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE53840_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE53840_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC
resid <- read.table("GSE53840_M_CTC_res.txt")
pheno <- read.delim("GSE53840 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#311219 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE53840_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3469 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE53840.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE53840"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE53840_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE53840_normalized_batch_corrected_beta.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE53840 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~hiv.viral.load,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE53840_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53840 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.2500448      
effect = summary(fit)$coefficients[,1][2]
effect
#effect:0.000164896 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr =0.0001425796 


#Entropy with CTC 

designCTC = model.matrix(~hiv.viral.load +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE53840_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53840 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.1407098       
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001176145 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 7.924723e-05   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7903582     

tiff('GSE53840 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7903582",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE53840 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~hiv.viral.load,
                      pheno)

M_all <- M_all[,pheno$Sample_Name]
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53840 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

M_none <- M_none[,pheno$Sample_Name]

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE53840 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE53840 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.4299374                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001800767             
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002272827         

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.19465                      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001450581           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.000111145         


write.table(pheno, "GSE53840 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE67751 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE67751"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE67751 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE67751 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

#DMPs
design=model.matrix(~age +
                      sex +
                      HIVstatus,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE67751_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#37 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE67751",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE67751_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE67751_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE67751_M_res.txt")
pheno <- read.delim("GSE67751 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#310730 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE67751_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2021 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE67751.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE67751"
setwd(path)
#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE67751_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE67751 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE67751 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

pheno <- pheno %>% select(-c(Bcell, CD4T, CD8T, Mono, NK)) 

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")

celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$geo_accession <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "geo_accession")

glimpse(pheno)

write.table(pheno,
            "GSE67751 Phenotypes.txt")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            HIVstatus,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE67751.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      HIVstatus +
                      CD4T +
                      Bcell +
                      Mono +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE67751_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#60 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE67751",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE67751_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE67751_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE67751_M_CTC_res.txt")
pheno <- read.delim("GSE67751 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#312375 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE67751_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#1823 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE67751.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE67751"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE67751_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE67751 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE67751 Phenotypes.txt", sep = "")

which(B==0)
B[B==0] <- min(B[B>0])
M <- logit2(B)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        HIVstatus,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$geo_accession <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "geo_accession")

#Save entropy plot
tiff('GSE67751_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67751 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.6995618 
effect = summary(fit)$coefficients[,1][2]
effect
#effect:8.661221e-05
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0002237196  

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           HIVstatus +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$geo_accession<- rownames(Entropy_CTC)
pheno$title
pheno <- left_join(pheno, Entropy_CTC, by = "geo_accession")

#Save entropy plot
tiff('GSE67751_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67751 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.4221127   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: -0.0001418791 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001759317   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8911446 

tiff('GSE67751 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.891",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE67751 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


cpg <- "cg13921251"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))



png('ConstantVMP GSE55763.png',
    width = 7,
    height = 5,
    unit = 'in',
    res = 600)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.2, colour = "deepskyblue4")+
  labs(y=paste("% methylation at",cpg)) +
  geom_smooth(method = "lm", colour = "slategrey") +
  xlim(19,105) +
  theme_minimal()
dev.off()

#### Correlation of age and cell types 

cor.test(pheno$age, pheno$CD8T, method = "pearson")
#-0.155

ggplot(pheno) +
  geom_jitter(aes(x = age, y = CD8T), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = CD8T), colour = "black") +
  labs(title = "Pearson's corr = -0.155")+
  theme_minimal()

cor.test(pheno$age, pheno$CD4T, method = "pearson")
#0.14

ggplot(pheno) +
  geom_jitter(aes(x = age, y = CD4T), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = CD8T), colour = "black") +
  labs(title = "Pearson's corr = 0.14")+
  theme_minimal()


cor.test(pheno$age, pheno$Bcell, method = "pearson")
#-0.228

ggplot(pheno) +
  geom_jitter(aes(x = age, y = Bcell), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = Bcell), colour = "black") +
  labs(title = "Pearson's corr = -0.228")+
  theme_minimal()


cor.test(pheno$age, pheno$NK, method = "pearson")
#-0.18

ggplot(pheno) +
  geom_jitter(aes(x = age, y = NK), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = NK), colour = "black") +
  labs(title = "Pearson's corr = -0.18")+
  theme_minimal()

cor.test(pheno$age, pheno$Mono, method = "pearson")
#-0.16

ggplot(pheno) +
  geom_jitter(aes(x = age, y = Mono), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = Mono), colour = "black") +
  labs(title = "Pearson's corr = -0.16")+
  theme_minimal()

cor.test(pheno$age, pheno$Gran, method = "pearson")
#0.2

ggplot(pheno) +
  geom_jitter(aes(x = age, y = Gran), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = age, y = Gran), colour = "black") +
  labs(title = "Pearson's corr = 0.2")+
  theme_minimal()


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex + HIVstatus,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$geo_accession <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "geo_accession")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67751 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$geo_accession <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "geo_accession")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE67751 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(geo_accession, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,geo_accession),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE67751 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2090369                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003330987           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002632608     

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.246199                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0002441772        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.000209193      


write.table(pheno, "GSE67751 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




#### GSE72775 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE72775"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE72775 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE72775 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

#DMPs
design=model.matrix(~age +
                      sex +
                      ethnicity,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE72775_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#11235 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE72775",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE72775_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72775_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE72775_M_res.txt")
pheno <- read.delim("GSE72775 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#175210 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE72775_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2485 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE72775.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE72775"
setwd(path)
#VMP check
cpg <- as.character(results_BP[4,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72775_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE72775 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE72775 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

pheno %>% transmute(celltypes = CD4T + CD8T + NK + Mono + Gran)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            ethnicity,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE72775.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      ethnicity +
                      CD4T +
                      Bcell +
                      CD8T +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE72775_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#16222 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE72775",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72775_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE72775_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE72775_M_CTC_res.txt")
pheno <- read.delim("GSE72775 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#195329 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE72775_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3129 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE72775.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE72775"
setwd(path)
#VMP check
cpg <- as.character(results_BP[4,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72775_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE72775 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE72775 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        ethnicity,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE72775_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72775 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.09070863 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001875676 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001105556  

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           ethnicity +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE72775_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72775 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.00390439   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002467705  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 8.491392e-05

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7790694 

tiff('GSE72775 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7790694",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE72775 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


cpg <- "cg13921251"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

png('ConstantVMP GSE55763.png',
    width = 7,
    height = 5,
    unit = 'in',
    res = 600)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.2, colour = "deepskyblue4")+
  labs(y=paste("% methylation at",cpg)) +
  geom_smooth(method = "lm", colour = "slategrey") +
  xlim(19,105) +
  theme_minimal()
dev.off()



#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design=model.matrix(~ sex +
                      ethnicity,
                    pheno)


fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72775 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72775 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE72775 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.009118068                
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003236837           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001234045       

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.9918559                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.403873e-06      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001374325       


write.table(pheno, "GSE72775 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE111629 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE111629"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE111629 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE111629 Phenotypes.txt")
glimpse(pheno)

ID <- colnames(M)
rownames(pheno) <- ID

pheno$ID == colnames(M)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE111629_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#18215 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE111629",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE111629_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE111629_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE111629_M_res.txt")
pheno <- read.delim("GSE111629 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#324045 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE111629_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#16883 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE111629.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE111629"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE111629_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg)) 
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE111629 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE111629 Phenotypes.txt")
glimpse(pheno)

pheno$ID <- colnames(M)

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$ID <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "ID")

write.table(pheno, 
            "GSE111629 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
ID <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(ID = ID)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE111629.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE111629_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#13233 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE111629",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE111629_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE111629_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE111629_M_CTC_res.txt")
pheno <- read.delim("GSE111629 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#346472 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE111629_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#12908 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE111629.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE111629"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE111629_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE111629 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE111629 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$ID <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "ID")

#Save entropy plot
tiff('GSE111629_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE111629 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.09432215   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 7.604693e-05   
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.53779e-05 


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$ID<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "ID")

#Save entropy plot
tiff('GSE111629_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE111629 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.02716773    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 7.683763e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 3.469136e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7666181    

tiff('GSE111629 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7666181",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE111629 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$ID <- rownames(Entropy_all)
pheno$ID <- as.character(pheno$ID)
pheno <- left_join(pheno, Entropy_all, by = "ID")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE111629 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$ID <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "ID")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE111629 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE111629 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.02389561             
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001401598         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.188202e-05    

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8094899              
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -1.174164e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.868056e-05    


write.table(pheno, "GSE111629 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE72774 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE72774"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE72774/GSE72774 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE72774/GSE72774 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

#DMPs
design=model.matrix(~age +
                      sex +
                      numberofyearsinschool.ch1 +
                      smokingstatus.ch1,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE72774_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#31180 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE72774",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE72774_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72774_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE72774_M_res.txt")
pheno <- read.delim("GSE72774/GSE72774 Phenotypes.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
resid <- as.data.frame(resid)
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#135239 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE72774_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#6686 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE72774.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE72774"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72774_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE72774/GSE72774 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE72774 Phenotypes.txt")
glimpse(pheno)

pheno %>% transmute(sumcelltypes = bcell.ch1 + CD4T + CD8T + Gran + Mono + nk.ch1)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex +
                                            smokingstatus.ch1 +
                                            numberofyearsinschool.ch1,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE72774.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      smokingstatus.ch1 +
                      numberofyearsinschool.ch1+
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE72774_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#33965 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"
#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE72774",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72774_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE72774_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- read.table("GSE72774_M_CTC_res.txt")
pheno <- read.delim("GSE72774/GSE72774 Phenotypes.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#142862 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE72774_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#6140 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE72774.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE72774"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72774_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy

B <- data.table::fread("GSE72774/GSE72774 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE72774 Phenotypes.txt", sep = "")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        numberofyearsinschool.ch1 +
                        smokingstatus.ch1,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

which(is.na(B_adj))
row.has.na <- apply(B_adj, 1, function(x){any(is.na(x))})
B_adj <- B_adj[!row.has.na,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE72774_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72774 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.2455719  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 9.692726e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 8.337658e-05   

#Entropy with CTC 

designCTC = model.matrix(~sex +numberofyearsinschool.ch1 +
                           smokingstatus.ch1 +
                           CD4T +
                           CD8T +
                           Mono +
                           nk.ch1 +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

which(is.na(B_CTC_adj))
row.has.na <- apply(B_CTC_adj, 1, function(x){any(is.na(x))})
B_CTC_adj <- B_CTC_adj[!row.has.na,]


Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE72774_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72774 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.02605122    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001602996   
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 7.181806e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.866033  

tiff('GSE72774 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.866033",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE72774 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design=model.matrix(~ sex +
                      numberofyearsinschool.ch1 +
                      smokingstatus.ch1,
                    pheno)


fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72774 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72774 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE72774 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.08520546               
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001575698          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 9.136424e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8923563                
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.386587e-05      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001024132      


write.table(pheno, "GSE72774 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE72776 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE72776"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE72776 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE72776 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE72776_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#1613 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE72776",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE72776_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72776_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs
resid <- read.table("GSE72776_M_res.txt")
pheno <- read.delim("GSE72776 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#347857 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE72776_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3078 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE72776.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE72776/GSE72776"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72776_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("./GSE72776/GSE72776 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("./GSE72776/GSE72776 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

pheno %>% transmute(celltypes = Bcell + CD4T + CD8T + NK + Mono + Gran)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE72776.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      CD4T +
                      CD8T +
                      Mono +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE72776_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#1131 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE72776",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72776_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE72776_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE72776_M_CTC_res.txt")
pheno <- read.delim("GSE72776 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#357055 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE72776_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#4553 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE72776.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE72776"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE72776_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE72776/GSE72776 beta after normalisation filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE72776 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE72776_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72776 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.2467337   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002169871  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001859917   

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE72776_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72776 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.6720531     
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect:5.853283e-05 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001377704 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =  8.738e-16
#cor = 0.7400905   

tiff('GSE72776 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7400905",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE72776 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


cpg <- "cg13921251"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

png('ConstantVMP GSE55763.png',
    width = 7,
    height = 5,
    unit = 'in',
    res = 600)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.2, colour = "deepskyblue4")+
  labs(y=paste("% methylation at",cpg)) +
  geom_smooth(method = "lm", colour = "slategrey") +
  xlim(19,105) +
  theme_minimal()
dev.off()


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design=model.matrix(~ sex,
                    pheno)


fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72775 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE72775 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE72775 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.01018563                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0005721392            
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002175155        

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2100965                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0002616958      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002071686        


write.table(pheno, "GSE72776 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE166611 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE166611"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE166611 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE166611 Phenotypes.txt")
glimpse(pheno)

#DMPs
design=model.matrix(~age +
                      bmi,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE166611_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#1 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE166611",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE166611_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE166611_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE166611_M_res.txt")
pheno <- read.delim("GSE166611 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE166611_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#1916 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE166611.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE166611"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE166611_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE166611 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE166611 Phenotypes.txt")
glimpse(pheno)

pheno <- pheno %>% select(-c(CD8T, CD4T, NK, Bcell, Mono, Gran))
celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)

pheno$Sample_Name <- paste(pheno$Sample_Name, pheno$Slide, pheno$Array, sep = "_")

pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE166611 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

pheno$Sample_Name == colnames(B)
#DMPs
design=model.matrix(~age +
                      bmi +
                      CD4T +
                      CD8T +
                      Mono +
                      Bcell +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE166611_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#2 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE166611",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE166611_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE166611_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- read.table("GSE166611_M_CTC_res.txt")
pheno <- read.delim("GSE166611 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#311219 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE166611_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#925 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE166611.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE166611"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE166611_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy

B <- data.table::fread("GSE166611 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE166611 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~bmi,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE166611_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE166611 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.3533241     
effect = summary(fit)$coefficients[,1][2]
effect
#effect:0.0001497593  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001588485 


#Entropy with CTC 

designCTC = model.matrix(~bmi +
                           CD4T +
                           Bcell +
                           Mono +
                           CD8T +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE166611_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE166611 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.6865235      
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 5.329313e-05 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.000130776   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.9957639    

tiff('GSE166611 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9957639 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE166611 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~bmi,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE166611 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE166611 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE166611 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.104708              
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003710362          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002217673     

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1738786               
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0001475054        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001058959     


write.table(pheno, "GSE166611 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




#### GSE164056 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE164056"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE164056 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE164056 Phenotypes.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE164056_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#9311 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE164056",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE164056_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE164056_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs
resid <- read.table("GSE164056_M_res.txt")
pheno <- read.delim("GSE164056 Phenotypes.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#652619 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE164056_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3683 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE164056.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE164056"
setwd(path)
#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE164056_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE164056 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE164056 Phenotypes.txt")
glimpse(pheno)

library(ChAMP)
celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

pheno %>% transmute(sumcelltypes = Bcell + CD4T + CD8T + Gran + Mono + NK)

write.table(pheno,
            "GSE164056 Phenotypes.txt")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE164056.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      CD4T +
                      CD8T +
                      Mono +
                      Bcell +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE164056_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#12452 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE164056",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE164056_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE164056_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE164056_M_CTC_res.txt")
pheno <- read.delim("GSE164056/GSE164056 Phenotypes.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#142862 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE164056_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#5868 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE164056.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE164056"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE164056_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy

B <- data.table::fread("GSE164056 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE164056 Phenotypes.txt", sep = "")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

which(is.na(B_adj))
#row.has.na <- apply(B_adj, 1, function(x){any(is.na(x))})
#B_adj <- B_adj[!row.has.na,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE164056_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE164056 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.01736241  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002800239  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001163204  

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           CD8T +
                           Mono +
                           Bcell +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

which(is.na(B_CTC_adj))
#row.has.na <- apply(B_CTC_adj, 1, function(x){any(is.na(x))})
#B_CTC_adj <- B_CTC_adj[!row.has.na,]


Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE164056_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE164056 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.06576607   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001308348  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 9.369214e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8061784   

tiff('GSE164056 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8061784",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE164056 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE164056 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE164056 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE164056 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.001561235             
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0004954444         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001535839    

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.891697              
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.600347e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001173247    


write.table(pheno, "GSE164056 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE85311 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE85311"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE85311 beta after normalisation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE85311 Phenotypes.txt")
glimpse(pheno)

champ.QC(beta = as.matrix(B),
         pheno = pheno$sex,
         dendrogram = FALSE)

colnames(B) == pheno$Sample_Name

pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

#DMPs
design=model.matrix(~age +
                      sex +
                      training.status,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE85311_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#8805 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE85311",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE85311_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[3]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE85311_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("methylation at",cpg))
dev.off()

#### VMPs
resid <- read.table("GSE85311_M_res.txt")
pheno <- read.delim("GSE85311 Phenotypes.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#367785 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE85311_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#5336 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE85311.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE85311"
setwd(path)
#VMP check
cpg <- "cg08422420"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE85311_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth, colour = training.status)) +
  geom_jitter()+
  labs(y=paste("methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE85311 beta after normalisation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE85311 Phenotypes.txt")
glimpse(pheno)

library(ChAMP)
celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

pheno %>% transmute(sumcelltypes = Bcell + CD4T + CD8T + Gran + Mono + NK)

write.table(pheno,
            "GSE85311 Phenotypes.txt")

#DMPs
design=model.matrix(~age +
                      sex +
                      training.status +
                      CD4T +
                      CD8T +
                      Mono +
                      Bcell +
                      NK,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE85311_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#578 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE85311",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE85311_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE85311_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- read.table("GSE85311_M_CTC_res.txt")
pheno <- read.delim("GSE85311/GSE85311 Phenotypes.txt", sep = "")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#142862 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE85311_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3776 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE85311.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE85311"
setwd(path)
#VMP check
cpg <- as.character(results_BP[5,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE85311_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("methylation at",cpg))
dev.off()

#### Entropy

B <- data.table::fread("GSE85311 beta after normalisation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE85311 Phenotypes.txt", sep = "")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        training.status,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

which(is.na(B_adj))
#row.has.na <- apply(B_adj, 1, function(x){any(is.na(x))})
#B_adj <- B_adj[!row.has.na,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE85311_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE85311 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.9905637   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 2.223144e-06  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001866718   

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           training.status +
                           CD4T +
                           CD8T +
                           Mono +
                           Bcell +
                           NK,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

which(is.na(B_CTC_adj))
#row.has.na <- apply(B_CTC_adj, 1, function(x){any(is.na(x))})
#B_CTC_adj <- B_CTC_adj[!row.has.na,]


Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE85311_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE85311 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.549951    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: -9.716753e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.000161003   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8613348    

tiff('GSE85311 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8061784",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE85311 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design=model.matrix(~ sex + training.status,
                    pheno)


fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE85311 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE85311 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE85311 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.06425977                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003064891             
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001605503          

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1166512                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0004076611       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002535778         


write.table(pheno, "GSE85311 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#### GSE151278 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE151278"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE151278 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE151278 Phenotypes.txt")
glimpse(pheno)

champ.SVD(beta = as.matrix(B),
          pd = pheno %>% select(age,
                                Sentrix_ID,
                                Sentrix_Position,
                                age.at.initiation.ch1,
                                drug,
                                disease.state,
                                sex,
                                response.ch1))

#DMPs
design=model.matrix(~age +
                      sex +
                      response.ch1,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE151278_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#152 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE151278",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE151278_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE151278_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE151278_M_res.txt")
pheno <- read.delim("GSE151278 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#369359 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE151278_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2912 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE151278.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE151278"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE151278_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE151278 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE151278 Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE151278 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            response.ch1,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE151278.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      response.ch1 +
                      CD4T +
                      NK +
                      Mono +
                      Bcell +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE151278_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#28 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE151278",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE151278_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE151278_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- data.table::fread("GSE151278_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE151278 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)
gc()

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)
gc()

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE151278_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2378 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE151278.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE151278"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE151278_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE151278 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE151278 Phenotypes.txt")

#run limma and adjust for covariates EXCLUDING age

design = model.matrix(~sex +
                        response.ch1,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE151278_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE151278 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.05822633     
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001772981    
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 9.203258e-05


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           response.ch1 +
                           CD4T +
                           NK +
                           Mono +
                           Bcell +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE151278_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE151278 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 4.047926e-05   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 9.463708e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 7.640418e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.822863    

tiff('GSE151278 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.822863",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE151278 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex +
                        response.ch1,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE151278 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE151278 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE151278 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.02134694                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002479607         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001052334      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.4872273                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect:8.212608e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001175697   


write.table(pheno, "GSE151278 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE96879 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE96879"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE96879 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE96879 Phenotypes.txt")
glimpse(pheno)

#DMPs
design=model.matrix(~age +
                      ethnicity +
                      disease.state,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE96879_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#22 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE96879",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE96879_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE96879_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE96879_M_res.txt")
pheno <- read.delim("GSE96879 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#367358 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE96879_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#1905 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE96879.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE96879"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE96879_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE96879 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE96879 Phenotypes.txt")
glimpse(pheno)


celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE96879 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ethnicity +
                                            disease.state,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE96879.rds')

#DMPs
design=model.matrix(~age +
                      ethnicity +
                      disease.state +
                      CD4T +
                      NK +
                      Mono +
                      Bcell +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE96879_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#11 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE96879",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE96879_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE96879_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- data.table::fread("GSE96879_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE96879 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)
gc()

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)
gc()

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE96879_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2090 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE96879.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE96879"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE96879_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy

B <- data.table::fread("GSE96879 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE96879 Phenotypes.txt")

#run limma and adjust for covariates EXCLUDING age

design = model.matrix(~ethnicity +
                        disease.state,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE96879_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE96879 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.01628075      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001682668     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 6.869033e-05



#Entropy with CTC 

designCTC = model.matrix(~ethnicity +
                           disease.state +
                           CD4T +
                           NK +
                           Mono +
                           Bcell +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE96879_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE96879 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.007405148    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001286493 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 4.692491e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.6916666     

tiff('GSE96879 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.6916666 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE96879 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~ethnicity +
                        disease.state,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE96879 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE96879 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE96879 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.03812135                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002283329               
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001084599           

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1831673                      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 8.74827e-05       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.520618e-05        


write.table(pheno, "GSE96879 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE134429 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE134429"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE134429 beta after normalisation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE134429 Phenotypes.txt")
glimpse(pheno)

#DMPs
design=model.matrix(~age +
                      sex +
                      batch +
                      patient.cohort +
                      donor,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE134429_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#492 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE134429",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE134429_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE134429_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE134429_M_res.txt")
pheno <- read.delim("GSE134429 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#650666 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE134429_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#10599 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE134429.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE134429"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE134429_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE134429 beta after normalisation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE134429 Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE134429 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            batch +
                                            patient.cohort +
                                            donor,
                                          pheno)) #Failed to run

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE134429.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      batch +
                      patient.cohort +
                      donor +
                      CD4T +
                      CD8T +
                      Mono +
                      Bcell +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE134429_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#164 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE134429",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE134429_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE134429_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- data.table::fread("GSE134429_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE134429 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#651544 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)
gc()

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)
gc()

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE134429_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#6391 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE134429.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE134429"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE134429_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE134429 beta after normalisation.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE134429 Phenotypes.txt")

#run limma and adjust for covariates EXCLUDING age

design = model.matrix(~sex + 
                        batch +
                        patient.cohort +
                        donor,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE134429_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE134429 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.2884682       
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.000124334     
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001161108 

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           batch +
                           patient.cohort + 
                           donor +
                           CD4T +
                           CD8T +
                           Mono +
                           Bcell +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE134429_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE134429 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.3170136     
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: -0.0001049645  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001040398  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8972672      

tiff('GSE134429 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8972672",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE134429 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related CpGs
#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

design = model.matrix(~sex + 
                        batch +
                        patient.cohort +
                        donor,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE134429 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#B_adj <- na.omit(B_adj)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE134429 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE134429 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.3510168                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -9.590444e-05             
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001020456          

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2711624                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.00015937       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001435181          


write.table(pheno, "GSE134429 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)






#### GSE120307 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE120307"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE120307 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE120307 Phenotypes.txt")

glimpse(pheno)
pheno$twin.pair<- as.factor(pheno$twin.pair)

colnames(B) == pheno$Sample_Name
#DMPs

design=model.matrix(~age + sex,
                    pheno)

Msubset <- M[sample(nrow(M),size=20000,replace=FALSE),]

corfit <- duplicateCorrelation(as.matrix(Msubset), #we fit limma models on M values and not beta values
                               design,
                               block=pheno$twin.pair) 
corfit$consensus
#0.2844905

fit1_M <- lmFit(M,
                design,
                block = pheno$twin.pair,
                correlation = 0.2844905)

fit2_M <- eBayes(fit1_M)

Bsubset <- B[sample(nrow(B),size=20000,replace=FALSE),]
corfitB <- duplicateCorrelation(Bsubset, #we fit limma models on M values and not beta values
                                design,
                                block=pheno$twin.pair) 
corfitB$consensus
#0.2818184

fit1_B <- lmFit(B,
                design,
                block = pheno$twin.pair,
                correlation = 0.2818184)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE120307_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE120307",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE120307_DMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE120307_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs
path = "/pvol/Preprocessing/Blood/GEO/GSE120307"
setwd(path)

resid <- data.table::fread("GSE120307_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE120307 Phenotypes.txt")
glimpse(pheno)

pheno$Sample_Name == colnames(resid)
design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- as.data.frame(resid)
resid <- resid[CpGs_to_keep,]

resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#chisq_pval["cg02734358",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)


#Save p-value histogram
tiff('GSE120307_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

results_BP %>% filter(PVALUE <0.005)
#915 VMPs 
#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE120307_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("Methylation at",cpg))
dev.off()

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE120307.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")
#### DMPs CTC 

create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 

path = "/pvol/Preprocessing/Blood/GEO/GSE120307"
setwd(path)

B <- data.table::fread("GSE120307 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE120307 Phenotypes.txt")


#use champ package to obtain cell type estimates 
library(ChAMP)
celltypes <- champ.refbase(beta = B, 
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno,
            file="GSE120307 Phenotypes.txt",
            quote=FALSE,
            row.names=TRUE,
            col.names=TRUE,
            sep='\t')

#DMPs with cell type correction (CTC)

glimpse(pheno)

design=model.matrix(~age +
                      sex + 
                      CD4T +
                      NK +
                      Bcell + 
                      Mono +
                      Gran,
                    pheno)

Msubset <- M[sample(nrow(M),size=10000,replace=FALSE),]

corfit <- duplicateCorrelation(Msubset, #we fit limma models on M values and not beta values
                               design,
                               block=pheno$twin.pair) 
corfit$consensus
#0.2945177

fit1_M <- lmFit(M,
                design,
                block = pheno$individual_id,
                correlation = 0.2945177)

fit2_M <- eBayes(fit1_M)

Bsubset <- B[sample(nrow(B),size=10000,replace=FALSE),]
corfitB <- duplicateCorrelation(Bsubset, #we fit limma models on M values and not beta values
                                design,
                                block=pheno$twin.pair) 
corfitB$consensus
#0.2777338

fit1_B <- lmFit(B,
                design,
                block = pheno$twin.pair,
                correlation = 0.2777338)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE120307_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`, binwidth = 30))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE120307",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE120307_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE120307_M_res_CTC.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#### VMPs CTC 
resid <- data.table::fread("GSE120307_M_res_CTC.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE120307 Phenotypes.txt")
rownames(pheno) <- pheno$Sample_Name

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- as.data.frame(resid)
resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

chisq_pval["cg13066289",]

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE120307_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 30))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

tiff('GSE120307_VMPCTCcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.02)+
  labs(y=paste("Methylation at",cpg))
dev.off()


results_BP %>% filter(PVALUE < 0.005)
#1208 VMPs

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE120307.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

####Entropy 
path = "/pvol/Preprocessing/Blood/GEO/GSE120307"
setwd(path)
B <- data.table::fread("GSE120307 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE120307 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)
fit1 <- lmFit(M,
              design,
              block = pheno$id,
              correlation = 0.2844905)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE120307_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE120307 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.3487482   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0003011363 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0003125076   

#Entropy with CTC 

designCTC = model.matrix(~ sex +
                           CD8T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  design,
                  block = pheno$twin.pair,
                  correlation = 0.2945177)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE120307_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE120307 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.9512259    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: -1.332493e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0002170295    

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.6509371   

tiff('GSE120307 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.6509371 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE120307 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)
fit1 <- lmFit(M_all,
              design,
              block = pheno$id,
              correlation = 0.2844905)
fit2 <- eBayes(fit1)


resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

B_adj[B_adj== 0] <- 0.000000000000000000000001
Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1120307 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)
fit1 <- lmFit(M_none,
              design,
              block = pheno$id,
              correlation = 0.2844905)
fit2 <- eBayes(fit1)


resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

B_adj[B_adj == 0] <- 0.00000000000000000000001
Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1120307 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE1120307 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8921025                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 3.685189e-05         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002695271      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.9041536                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -2.646002e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0002180036  


write.table(pheno, "GSE1120307 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE20236 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE20236"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE20236/GSE20236 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE20236/GSE20236 Phenotypes.txt")
glimpse(pheno)

pheno$Sample_Name <- pheno$geo_accession
pheno$Sample_Name == colnames(B)

#DMPs
design=model.matrix(~age,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE20236_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#0 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE20236",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE20236_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE20236_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

write.table(pheno,
            "GSE20236 Phenotypes.txt")
#### VMPs
resid <- read.table("GSE20236_M_res.txt")
pheno <- read.delim("GSE20236 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#353581 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE20236_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#74 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE20236.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE20236"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE20236_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE20236/GSE20236 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE20236 Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B)
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE20236 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#DMPs
design=model.matrix(~age +
                      CD4T +
                      CD8T +
                      Bcell +
                      NK +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE20236_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#1 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE20236",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE20236_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE20236_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE20236_M_CTC_res.txt")
pheno <- read.delim("GSE20236 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE20236_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#74 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE20236.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE20236"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE20236_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE20236/GSE20236 beta after filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE20236 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
#design = model.matrix(~sex +
#                        disease.status +
#                        smoking.status,
#                      pheno)

#fit1 <- lmFit(M,
#              design)
#fit2 <- eBayes(fit1)

#resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

#M_adj <- resid_M + rowMeans(M)
#plotDensities(M_adj, legend = F)

#B_adj <- ilogit2(M_adj)
#plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE20236_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE20236 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.2295936   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 5.892163e-06   
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.871463e-06


#Entropy with CTC 

designCTC = model.matrix(~ CD4T +
                           CD8T +
                           Bcell +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE20236_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE20236 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.03455883    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 8.244211e-06
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 3.8422e-06  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.4031118   

tiff('GSE20236 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.4031118",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE20236 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_all, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE20236 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_none, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE20236 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE20236 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.09465826                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 8.238226e-05          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.877801e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1785174                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -6.901783e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.090555e-05     


write.table(pheno, "GSE20236 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE19711 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE19711"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE19711/GSE19711 beta after filtering and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE19711/GSE19711 Phenotypes after preprocessing.txt")
glimpse(pheno)

pheno$Sample_Name <- pheno$ID
colnames(B) == pheno$Sample_Name

#DMPs
design=model.matrix(~age +
                      ca125.ch1,
                    pheno)
keep <- pheno %>% filter(!is.na(ca125.ch1))
B <- B[,keep$Sample_Name] 
M <- M[,keep$Sample_Name]

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE19711_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#32 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE19711",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE19711_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(keep, meth= as.numeric(B[cpg,]))
tiff('GSE19711_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE19711_M_res.txt")
pheno <- read.delim("GSE19711/GSE19711 Phenotypes after preprocessing.txt")

design2 = model.matrix(~ age, keep)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE19711_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#97 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE19711.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE19711"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(keep, meth= as.numeric(B[cpg,]))
tiff('GSE19711_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE19711/GSE19711 beta after filtering and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE19711/GSE19711 Phenotypes after preprocessing.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B)
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno$Sample_Name <- pheno$ID
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE19711 Phenotypes after preprocessing.txt",
            col.names = TRUE,
            sep = "\t")

#DMPs
design=model.matrix(~age +
                      ca125.ch1 +
                      CD4T +
                      CD8T +
                      Bcell +
                      NK +
                      Gran,
                    pheno)
B <- B[,keep$Sample_Name]
M <- M[,keep$Sample_Name]

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE19711_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#27 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE19711",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(keep, meth= as.numeric(B[cpg,]))
tiff('GSE19711_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE19711_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE19711_M_CTC_res.txt")
pheno <- read.delim("GSE19711 Phenotypes after preprocessing.txt")

design2 = model.matrix(~ age, keep)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE19711_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#103 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE19711.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE19711"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(keep, meth= as.numeric(B[cpg,]))
tiff('GSE19711_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE19711/GSE19711 beta after filtering and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE19711 Phenotypes after preprocessing.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~ca125.ch1,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- pheno %>% filter(!is.na(ca125.ch1))
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE19711_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE19711 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.3314212      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001135034  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001166443 


#Entropy with CTC 

designCTC = model.matrix(~ca125.ch1 +
                           CD4T +
                           Bcell +
                           CD8T +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE19711_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE19711 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.7376015       
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 2.904989e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 8.66153e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7472739      

tiff('GSE19711 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7472739",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE19711 Phenotypes after preprocessing.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~ca125.ch1,
                      pheno)

M_all <- M_all[,pheno$Sample_Name]
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE19711 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

M_none <- M_none[,pheno$Sample_Name]

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE19711 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE19711 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.1802643              
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002042859          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.000152049     

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8299482               
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 2.314042e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001076378     


write.table(pheno, "GSE19711 Phenotypes after preprocessing.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE157131 ####
#450K analysis

path = "/pvol/Preprocessing/Blood/GEO/GSE157131"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE157131 450K beta after normalisation and filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE157131 450K Phenotypes.txt")
glimpse(pheno)


champ.SVD(beta = as.matrix(B),
          pd = pheno %>% select(age,
                                sex,
                                diastolic.blood.pressure..mmhg..ch1,
                                systolic.blood.pressure..mmhg..ch1))

#DMPs
design=model.matrix(~age,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE157131_450K_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#1853 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE157131_450K",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE157131_450K_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE157131_450K_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE157131_450K_M_res.txt")
pheno <- read.delim("GSE157131 450K Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,] 
#650666 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE157131_450K_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#5078 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE157131_450K.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE157131"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE157131_450K_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE157131 450K beta after normalisation and filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE157131 450K Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "450K")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE157131 450K Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1)

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE157131 450K.rds')

#DMPs
design=model.matrix(~age +
                      CD4T +
                      NK +
                      Mono +
                      Bcell +
                      Gran,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE157131_450K_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#1447 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE157131_450K",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE157131_450K_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE157131_450K_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- data.table::fread("GSE157131_450K_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE157131 450K Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)
gc()

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)
gc()

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE157131_450K_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#6759 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE157131_450K.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE157131"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE157131_450K_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE157131 450K beta after normalisation and filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE157131 450K Phenotypes.txt")

#run limma and adjust for covariates EXCLUDING age

#design = model.matrix(~sex + 
#                        batch +
#                        patient.cohort +
#                        donor,
#                      pheno)

#fit1 <- lmFit(M,
#design)
#fit2 <- eBayes(fit1)

#resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

#M_adj <- resid_M + rowMeans(M)
#B_adj <- ilogit2(M_adj)
#plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE157131_450K_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE157131 450K Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.1403107         
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002125616       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 0.0001437215  

#Entropy with CTC 

designCTC = model.matrix(~ CD4T +
                           NK +
                           Mono +
                           Bcell +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE157131_450K_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE157131 450K Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.5701558      
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 7.182903e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 0.0001263443   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8786168       

tiff('GSE157131 450K Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8786168",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE157131 450K Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_all, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE157131 450K Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_none, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE157131 450K Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE157131 450K Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.029841                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0004075167           
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001866128      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.7411363                     
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -5.068415e-05         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001532686   


write.table(pheno, "GSE157131 450K Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#EPIC analysis

path = "/pvol/Preprocessing/Blood/GEO/GSE157131"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE157131 EPIC beta after normalisation and filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE157131 EPIC Phenotypes.txt")
glimpse(pheno)

#DMPs
design=model.matrix(~age,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE157131_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#154502 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE157131_EPIC",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE157131_EPIC_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE157131_EPIC_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE157131_EPIC_M_res.txt")
pheno <- read.delim("GSE157131 EPIC Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,] 
#650666 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE157131_EPIC_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#48285 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE157131_EPIC.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE157131"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE157131_EPIC_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE157131 EPIC beta after normalisation and filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE157131 EPIC Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE157131 EPIC Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)
#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1)

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE157131 EPIC.rds')

#DMPs
design=model.matrix(~age +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)


coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE157131_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#132035 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE157131_EPIC",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE157131_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE157131_EPIC_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- data.table::fread("GSE157131_EPIC_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid %>% select(-V1)
pheno <- read.delim("GSE157131 EPIC Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)
gc()

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)
gc()

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE157131_EPIC_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#64063 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE157131_EPIC.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE157131"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE157131_EPIC_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE157131 EPIC beta after normalisation and filtering.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE157131 EPIC Phenotypes.txt")

#run limma and adjust for covariates EXCLUDING age

#design = model.matrix(~sex + 
#                        batch +
#                        patient.cohort +
#                        donor,
#                      pheno)

#fit1 <- lmFit(M,
design)
#fit2 <- eBayes(fit1)

#resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

#M_adj <- resid_M + rowMeans(M)
#B_adj <- ilogit2(M_adj)
#plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE157131_EPIC_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE157131 EPIC Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.0003969457        
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002206612      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 6.206786e-05  

#Entropy with CTC 

designCTC = model.matrix(~ CD4T +
                           NK +
                           Mono +
                           Bcell +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

row.has.na <- lapply(B_CTC_adj, function(x){any(is.na(x))})

B_CTC_adj <-na.omit(B_CTC_adj)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE157131_EPIC_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE157131 EPIC Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.002077936      
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001462794  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 4.737907e-05   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8972672      

tiff('GSE157131 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8972672",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE157131 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_all, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE157131 EPIC Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_none, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE157131 EPIC Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE157131 EPIC Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 6.54807e-07                     
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0004215814            
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 8.417384e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.7718038                      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -1.7462e-05         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.019273e-05   


write.table(pheno, "GSE157131 EPIC Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)




#### GSE117589 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE117859"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE117859 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE117859 Phenotypes.txt")
glimpse(pheno)

#DMPs
design=model.matrix(~age +
                      smoking,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE117859_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#6846 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE117859",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE117859_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE117859_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs
resid <- read.table("GSE117859_M_res.txt")
pheno <- read.delim("GSE117859 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#344620 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE117859_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#8083 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE117859.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE117859"
setwd(path)
#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE117859_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg)) 
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE117859 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE117859 Phenotypes.txt")
glimpse(pheno)

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ smoking,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE117859.rds')

#DMPs
design=model.matrix(~age +
                      smoking +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE117859_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#14648 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"

create_summary(toptable = results,
               dataset_label = "GSE117859",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE117859_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE117859_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE117859_M_CTC_res.txt")
pheno <- read.delim("GSE117859 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#339840 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE117859_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#5827 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE117859.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE117859"
setwd(path)
#VMP check
cpg <- as.character(results_BP[10,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE117859_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE117859 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE117859 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~smoking,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE117859_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE117859 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.6654736   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 3.190815e-05    
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 7.376238e-05  


#Entropy with CTC 

designCTC = model.matrix(~smoking +
                           Bcell +
                           CD8T +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE117859_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE117859 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.5534623    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 3.250113e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 5.481625e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7481743   

tiff('GSE117859 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr =  ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE117859 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~smoking,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1117859 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1117859 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE1117859 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.6654736                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 3.190815e-05         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 7.376238e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.02726577                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0001427769        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.45158e-05  


write.table(pheno, "GSE1117859 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE117860 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE117860"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE117860 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE117860 Phenotypes.txt")
glimpse(pheno)


#DMPs
design=model.matrix(~age +
                      smoking,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE117860_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#3238 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE117860",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE117860_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE117860_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE117860_M_res.txt")
pheno <- read.delim("GSE117860 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#310216 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE117860_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#7418 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE117860.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE117860"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE117860_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE117860 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE117860 Phenotypes.txt")
glimpse(pheno)

#celltypes <- champ.refbase(beta = B,
#                           arraytype = "450K")
#celltypes <- celltypes[[2]]
#celltypes <- as.data.frame(celltypes)
#celltypes$Sample_Name <- rownames(celltypes)
#pheno <- left_join(pheno, celltypes, by = "Sample_Name")

#write.table(pheno, 
#            "GSE117860 Phenotypes.txt",
#            col.names = TRUE,
#            sep = "\t")


#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ smoking,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE117860.rds')

#DMPs
design=model.matrix(~age +
                      smoking +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE117860_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#6539 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE117860",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE117860_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE117860_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE117860_M_CTC_res.txt")
pheno <- read.delim("GSE117860 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#311219 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE117860_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#8542 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE117860.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE117860"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE117860_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE117860 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE117860 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~smoking,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE117860_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE117860 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.1489717   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 9.180556e-05   
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 6.352066e-05


#Entropy with CTC 

designCTC = model.matrix(~smoking +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE117860_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE117860 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.07426908    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 8.187646e-05
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 4.57793e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7258579   

tiff('GSE117860 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7258579",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE117860 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~smoking,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1117860 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1117860 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE1117860 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.09585561                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001638837         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 9.823557e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.8964397                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -6.760481e-06        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.191463e-05  


write.table(pheno, "GSE1117860 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE147740 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE147740"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE147740 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE147740 Phenotypes.txt")
pheno_replicates <- read.delim("GSE147740 Phenotypes with average replicates.txt")
glimpse(pheno)
pheno$sex <- as.factor(pheno$sex)
pheno$age <- as.numeric(pheno$age)

pheno$Sample_Name <- paste0(pheno$geo_accession,"_",pheno$title)

write.table(pheno,
            "GSE147740 Phenotypes.txt")

#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE147740_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#70852 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE147740",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE147740_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE147740_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### VMPs
resid <- read.table("GSE147740_M_res.txt")
pheno <- read.delim("GSE147740 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#435116 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg16699354",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE147740_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#15220 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE147740.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE147740"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE147740_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE147740 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE147740 Phenotypes.txt", sep = "")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE147740.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE147740_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#131778 DMPs 
#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE147740",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE147740_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("% methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE147740_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE147740_M_CTC_res.txt")
pheno <- read.delim("GSE147740 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#455462 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg00231274",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE147740_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#9746 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE147740.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE147740"
setwd(path)
#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE147740_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("% methylation at",cpg))
dev.off()

#### Entropy

B <- data.table::fread("GSE147740 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE147740 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE147740_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE147740 Entropy",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.8578721 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -1.341886e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 7.491189e-05

#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           CD8T +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE147740_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE147740 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.4528796   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 4.863148e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 6.47635e-05 

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.865537  

tiff('GSE147740 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.865537",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE147740 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


cpg <- "cg13921251"
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))

png('ConstantVMP GSE55763.png',
    width = 7,
    height = 5,
    unit = 'in',
    res = 600)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.2, colour = "deepskyblue4")+
  labs(y=paste("% methylation at",cpg)) +
  geom_smooth(method = "lm", colour = "slategrey") +
  xlim(19,105) +
  theme_minimal()
dev.off()


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE147740 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE147740 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE147740 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.4859705                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 5.396395e-05         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 7.742493e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.3093811                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -9.264769e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 9.109677e-05  


write.table(pheno, "GSE147740 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE152026 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE152026"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE152026 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE152026 Phenotypes.txt")
glimpse(pheno)

#some ages are missing 

pheno <- pheno %>% filter(!is.na(age))
B <- B[,pheno$Sample_Name]
M <- M[,pheno$Sample_Name]

#DMPs
design=model.matrix(~age +
                      sex +
                      disease.status,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE152026_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#152495 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE152026",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE152026_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE152026_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE152026_M_res.txt")
pheno <- read.delim("GSE152026 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE152026_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#36013 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE152026.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE152026"
setwd(path)
#VMP check
cpg <- as.character(results_BP[4,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE152026_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE152026 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE152026 Phenotypes.txt")
glimpse(pheno)

B <- B[,pheno$Sample_Name]
M <- M[,pheno$Sample_Name]

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE152026 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex +
                                            disease.status,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE152026.rds')


#DMPs
design=model.matrix(~age +
                      sex +
                      disease.status +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE152026_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#163911 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE152026",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE152026_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE152026_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE152026_M_CTC_res.txt")
pheno <- read.delim("GSE152026 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE152026_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#29509 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE152026.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE152026"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE152026_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE152026 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE152026 Phenotypes.txt")


#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex + disease.status,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE152026_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE152026 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.0006154573      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001244897 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 3.622545e-05


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           disease.status +
                           CD4T +
                           Bcell +
                           Mono +
                           CD8T +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE152026_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE152026 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 5.892645e-05       
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001302386
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 3.227122e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8915562      

tiff('GSE152026 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8915562 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE152026 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex + disease.status,
                      pheno)
M_all <- M_all[,pheno$Sample_Name]
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE152026 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

M_none <- M_none[,pheno$Sample_Name]
fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE152026 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE152026 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 2.078894e-11                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002943934          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.339288e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.06430396                    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -9.196096e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.964742e-05  


write.table(pheno, "GSE152026 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE132203 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE132203"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE132203 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE132203 Phenotypes.txt")
glimpse(pheno)

#some ages are missing 

#pheno <- pheno %>% filter(!is.na(age))
#B <- B[,pheno$Sample_Name]
#M <- M[,pheno$Sample_Name]


library(mice)
md.pattern(pheno)
glimpse(pheno)

pheno$mergedcapsandpsswinthin30days.ch1 <- as.factor(pheno$mergedcapsandpsswinthin30days.ch1)

pheno$childabphyssexemot_ctq_01modandsev.ch1 <- as.factor(pheno$childabphyssexemot_ctq_01modandsev.ch1)
pheno$ethnicity <- as.factor(pheno$ethnicity)
pheno$sex <- as.factor(pheno$sex)

pheno_imputations <- pheno %>% select(age,
                                      childabphyssexemot_ctq_01modandsev.ch1,
                                      sex,
                                      mergedcapsandpsswinthin30days.ch1 ,
                                      ethnicity,
                                      tei_total_types_experienced_somewitness.ch1)

imputed <- mice(data = pheno_imputations,
                m = 10)

imputed$imp
pheno_imputed <- complete(imputed)
plot(imputed)

pheno_imputations <- pheno %>% select(title,
                                      geo_accession,
                                      Slide,
                                      Array,
                                      Bcell,
                                      CD8T,
                                      CD4T,
                                      Neutro,
                                      Mono,
                                      NK,
                                      Sample_Name) %>% cbind(pheno_imputed)

write.table(pheno_imputations,
            "GSE132203 Phenotypes with imputations.txt")

#DMPs
design=model.matrix(~age +
                      sex +
                      ethnicity + 
                      childabphyssexemot_ctq_01modandsev.ch1,
                    pheno_imputations)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE132203_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#145682 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE132203",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE132203_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE132203_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE132203_M_res.txt")
pheno <- read.delim("GSE132203 Phenotypes with imputations.txt")

design2 = model.matrix(~ age, pheno_imputations)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE132203_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#14416 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE132203.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE132203"
setwd(path)
#VMP check
cpg <- as.character(results_BP[2,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE132203_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE132203 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE132203 Phenotypes with imputations.txt", sep = " ") %>%
  mutate(Sample_Name = paste(geo_accession,
                             Sample_Name,
                             sep="_"))
glimpse(pheno)

#celltypes <- champ.refbase(beta = B,
#                           arraytype = "EPIC")
#celltypes <- celltypes[[2]]
#celltypes <- as.data.frame(celltypes)
#celltypes$Sample_Name <- rownames(celltypes)
#pheno <- left_join(pheno, celltypes, by = "Sample_Name")

#write.table(pheno, 
#            "GSE132203 Phenotypes.txt",
#            col.names = TRUE,
#            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex +
                                            ethnicity + 
                                            childabphyssexemot_ctq_01modandsev.ch1,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE132203.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      ethnicity + 
                      childabphyssexemot_ctq_01modandsev.ch1 +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE132203_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#171033 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE132203",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE132203_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE132203_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE132203_M_CTC_res.txt")
pheno <- read.delim("GSE132203 Phenotypes with imputations.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE132203_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#11923 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE132203.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE132203"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE132203_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy

B <- data.table::fread("GSE132203 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno_imputations <- read.delim("GSE132203 Phenotypes with imputations.txt", sep = " ")
pheno <- read.delim("GSE132203 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex + ethnicity + childabphyssexemot_ctq_01modandsev.ch1,
                      pheno_imputations)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- pheno %>% select(-Sample_Name)

pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE132203_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE132203 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#8.252735e-08      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0002296967  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 4.244052e-05 


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           ethnicity +
                           childabphyssexemot_ctq_01modandsev.ch1 +
                           CD4T +
                           Bcell +
                           Mono +
                           CD8T +
                           Neutro,
                         pheno_imputations)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE132203_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE132203 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 7.738099e-10        
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0002213623 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 3.555288e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8915562      

tiff('GSE132203 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8915562 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE132203 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex + ethnicity + childabphyssexemot_ctq_01modandsev.ch1,
                      pheno_imputations)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE132203 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE132203 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE132203 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 3.119036e-10                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.000355602 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 5.578792e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.162107                   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 6.882591e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr:4.918513e-05  


write.table(pheno, "GSE132203 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE100264 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE100264"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE100264 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE100264 Phenotypes.txt")
glimpse(pheno)

#DMPs
design=model.matrix(~age +
                      hcv_dx,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE100264_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#1133 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE100264",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE100264_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE100264_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE100264_M_res.txt")
pheno <- read.delim("GSE100264 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE100264_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3530 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE100264.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE100264"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE100264_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE100264 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE100264 Phenotypes.txt")
glimpse(pheno)

#celltypes <- champ.refbase(beta = B,
#                           arraytype = "450K")
#celltypes <- celltypes[[2]]
#celltypes <- as.data.frame(celltypes)
#celltypes$Sample_Name <- rownames(celltypes)
#pheno <- left_join(pheno, celltypes, by = "Sample_Name")

#write.table(pheno, 
#            "GSE100264 Phenotypes.txt",
#            col.names = TRUE,
#            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~hcv_dx,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE100264.rds')

#DMPs
design=model.matrix(~age +
                      hcv_dx +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE100264_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#2677 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"
#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE100264",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE100264_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE100264_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE100264_M_CTC_res.txt")
pheno <- read.delim("GSE100264 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#311219 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE100264_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#2152 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE100264.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE100264"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE100264_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE100264 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE100264 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~hcu_dx,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE100264_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE100264 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.9039949    
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -1.177589e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 9.756643e-05


#Entropy with CTC 

designCTC = model.matrix(~hcv_dx +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE100264_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE100264 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.7386943     
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: -3.318863e-05 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 9.941942e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.9957639    

tiff('GSE100264 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.9957639 ",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE100264 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)



#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~hcv_dx,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE100264 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE100264 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE100264 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2185941          
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001758535      
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001427044  

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.08593418           
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -0.0002119981       
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.000123135   


write.table(pheno, "GSE100264 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)





#### GSE107080 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE107080"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE107080 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE107080 Phenotypes.txt")
glimpse(pheno)

#multiple imoutations for phenotypes 

library(mice)
md.pattern(pheno)
glimpse(pheno)

pheno$hcv_dx <- as.factor(pheno$hcv_dx)
pheno$race <- as.factor(pheno$race)
pheno$Smoking <- as.factor(pheno$Smoking)
pheno$idu <- as.factor(pheno$idu)
pheno$sex <- as.factor(pheno$sex)
pheno$HIV <- as.factor(pheno$HIV)

pheno_imputations <- pheno %>% select(age,
                                      artadherence,
                                      hcv_dx,
                                      HIV,
                                      idu,
                                      race,
                                      Smoking,
                                      sex,
                                      wbc_new.ch1)

imputed <- mice(data = pheno_imputations,
                m = 10)

imputed$imp
pheno_imputed <- complete(imputed)
plot(imputed)

pheno_imputations <- pheno %>% select(title,
                                      geo_accession,
                                      Sentrix_ID,
                                      Sentrix_Position,
                                      description.2,
                                      Bcell,
                                      CD8T,
                                      CD4T,
                                      NK,
                                      Mono,
                                      Gran,
                                      Sample_Name, 
                                      wbc_new.ch1) %>% cbind(pheno_imputed)

write.table(pheno_imputations,
            "GSE107080 Phenotypes with imputations.txt")

#DMPs
design=model.matrix(~age +
                      hcv_dx +
                      race +
                      Smoking +
                      artadherence,
                    pheno_imputations)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE107080_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#832 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE107080",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE107080_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE107080_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE107080_M_res.txt")
pheno <- read.delim("GSE107080 Phenotypes with imputations.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE107080_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3930 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE107080.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE107080"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE107080_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE107080 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE107080 Phenotypes with imputations.txt")
glimpse(pheno)

#celltypes <- champ.refbase(beta = B,
#                           arraytype = "450K")
#celltypes <- celltypes[[2]]
#celltypes <- as.data.frame(celltypes)
#celltypes$Sample_Name <- rownames(celltypes)
#pheno <- left_join(pheno, celltypes, by = "Sample_Name")

#write.table(pheno, 
#            "GSE107080 Phenotypes.txt",
#            col.names = TRUE,
#            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~hcv_dx +
                                            Smoking +
                                            race +
                                            artadherence,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE107080.rds')

#DMPs
design=model.matrix(~age +
                      hcv_dx +
                      Smoking +
                      race +
                      artadherence +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE107080_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#2834 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE107080",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE107080_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE107080_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE107080_M_CTC_res.txt")
pheno <- read.delim("GSE107080 Phenotypes with imputations.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE107080_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#4247 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE107080.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE107080"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE107080_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE107080 beta normalised and batch corrected.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE107080 Phenotypes with imputations.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~hcv_dx +
                        race +
                        Smoking +
                        artadherence,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE107080_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE107080 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.376273     
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 7.605117e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 8.585942e-05


#Entropy with CTC 

designCTC = model.matrix(~hcv_dx +
                           race +
                           Smoking +
                           artadherence +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE107080_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE107080 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.1803641      
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 8.5567e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 6.376264e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7466007     

tiff('GSE107080 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7466007",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE107080 Phenotypes with imputations.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~hcv_dx +
                        race +
                        Smoking +
                        artadherence,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE107080 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE107080 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE107080 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.2098926            
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001584051        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001261323   

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.6157777             
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -3.80636e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 7.578868e-05   


write.table(pheno, "GSE107080 Phenotypes with imputations.txt",
            sep = "\t",
            col.names = TRUE)


#### GSE116339 ####
path = "/pvol/Preprocessing/Blood/GEO/GSE116339"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE116339 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE116339 Phenotypes.txt")
glimpse(pheno)

pheno$Sample_Name == colnames(B)

#DMPs
design=model.matrix(~age +
                      sex +
                      pbb.77.ch1,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE116339_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#82177 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE116339",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE116339_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE116339_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE116339_M_res.txt")
pheno <- read.delim("GSE116339 Phenotypes with imputations.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE116339_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#67333 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE116339.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE116339"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE116339_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE116339 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE116339 Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE116339 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex +
                                            pbb.77.ch1,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE116339.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      pbb.77.ch1 +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE116339_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#91808 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE116339",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE116339_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE116339_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- read.table("GSE116339_M_CTC_res.txt")
pheno <- read.delim("GSE116339 Phenotypes with imputations.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE116339_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#58574 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE116339.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE116339"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE116339_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE116339 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE116339 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        pbb.77.ch1,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
plotDensities(M_adj, legend = F)

B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE116339_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE116339 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.5539667      
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.750642e-05 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 2.956551e-05


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           pbb.77.ch1 +
                           CD4T +
                           Bcell +
                           Mono +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE116339_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE116339 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.001710191       
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 6.457464e-05  
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 2.050586e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.7022475      

tiff('GSE116339 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.7022475",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE116339 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)


#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex +
                        pbb.77.ch1,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1116339 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE1116339 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE1116339 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.7313081                
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.562418e-05         
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 4.548148e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.4511984                 
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 1.987502e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 2.636444e-05      


write.table(pheno, "GSE1116339 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE168739 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE168739"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE168739 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE168739 Phenotypes.txt")
glimpse(pheno)

pheno$Sample_Name == colnames(B)
#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE168739_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#94461 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE168739",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE168739_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE168739_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs 
resid <- read.table("GSE168739_M_res.txt")
pheno <- read.delim("GSE168739 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#324045 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE168739_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#22511 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE168739.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE168739"
setwd(path)
#VMP check
cpg <- as.character(results_BP[15000,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE168739_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg)) 
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE168739 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE168739 Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE168739 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~ sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE168739.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE168739_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#103871 DMPs 
directory = "/pvol/prelim_data_ARC_2023/"
#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE168739",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE168739_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE168739_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC 
resid <- read.table("GSE168739_M_CTC_res.txt")
pheno <- read.delim("GSE168739 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#346472 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE168739_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#16294 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE168739.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE168739"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE168739_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE168739 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE168739 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE168739_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE168739 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#0.005971567   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0001940455  
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 7.020156e-05  


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           Bcell +
                           CD8T +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name <- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE168739_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE168739 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 0.0008300285    
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0001549404 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 4.600109e-05

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.6611316    

tiff('GSE168739 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.6611316",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE168739 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE168739 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE168739 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE168739 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.001160097              
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.000368954          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 0.0001127567      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.7208878               
effect = summary(fit)$coefficients[,1][2]
effect
#effect: -2.155022e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 6.027622e-05    


write.table(pheno, "GSE168739 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



#### GSE197674 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE197674"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)

#### DMPs
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE197674 beta after filtering and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE197674 Phenotypes.txt")
glimpse(pheno)

pheno$Sample_Name == colnames()

#DMPs
design=model.matrix(~age +
                      sex +
                      abdominal_pelvic_rt.ch1 +
                      alkylating.agent.classic.ch1+
                      anthracyclines.ch1 +
                      corticosteroids.ch1 +
                      epipodophyllotoxins.ch1 +
                      platinum.ch1 +
                      vincristine.ch1 +
                      chestrt.ch1 +
                      brainrt.ch1,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE197674_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#677964 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE197674",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE197674_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE197674_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))+
  theme(legend.position = "none")
dev.off()

#### VMPs 
resid <- data.table::fread("GSE197674_M_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid[,-1]
pheno <- read.delim("GSE197674 Phenotypes.txt")
glimpse(pheno)

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,] 
#324045 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

#TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE197674_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#117817 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE197674.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE197674"
setwd(path)
#VMP check
cpg <- as.character(results_BP[3,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE197674_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg)) 
dev.off()

#### DMPs CTC
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE197674 beta after filtering and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE197674 Phenotypes.txt")
#glimpse(pheno)

library(ChAMP)
celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")

celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE197674 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF
colnames(frac.m) <- paste(colnames(frac.m),
                          "EpiDISH")
Sample_Name <- rownames(frac.m)
tomerge <- frac.m %>%
  as_tibble() %>%
  mutate(Sample_Name = Sample_Name)
tomerge <- left_join(pheno,
                     tomerge)

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex +
                                            abdominal_pelvic_rt.ch1 +
                                            alkylating.agent.classic.ch1+
                                            anthracyclines.ch1 +
                                            corticosteroids.ch1 +
                                            epipodophyllotoxins.ch1 +
                                            platinum.ch1 +
                                            vincristine.ch1 +
                                            chestrt.ch1 +
                                            brainrt.ch1,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE197674.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      abdominal_pelvic_rt.ch1 +
                      alkylating.agent.classic.ch1+
                      anthracyclines.ch1 +
                      corticosteroids.ch1 +
                      epipodophyllotoxins.ch1 +
                      platinum.ch1 +
                      vincristine.ch1 +
                      chestrt.ch1 +
                      brainrt.ch1 +
                      `B EpiDISH`+
                      `NK EpiDISH` +
                      `CD4T EpiDISH` +
                      `CD8T EpiDISH` +
                      `Mono EpiDISH`,
                    #`Neutro EpiDISH`,
                    tomerge)
# CD4T +
# Bcell +
# CD8T +
# NK +
# Gran,
# pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE197674_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#376368 DMPs 

#directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
directory = "/pvol/prelim_data_ARC_2023/"
create_summary(toptable = results,
               dataset_label = "GSE197674",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE197674_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE197674_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- data.table::fread("GSE197674_M_CTC_res.txt")
resid <- as.data.frame(resid)
rownames(resid) <- resid$V1
resid <- resid[,-1]
pheno <- read.delim("GSE197674 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
#shapirotest <- apply(as.matrix(resid),1,shapiro.test)
#pvals <- sapply(shapirotest,"[[",2)
#CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
#resid <- resid[CpGs_to_keep,]
#346472 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE197674_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#67964 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE197674.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE197674"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE197674_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE197674 beta after filtering and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE197674 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        abdominal_pelvic_rt.ch1 +
                        alkylating.agent.classic.ch1+
                        anthracyclines.ch1 +
                        corticosteroids.ch1 +
                        epipodophyllotoxins.ch1 +
                        platinum.ch1 +
                        vincristine.ch1 +
                        chestrt.ch1 +
                        brainrt.ch1,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE197674_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197674 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval4.399355e-28 
#0.1096534  
effect = summary(fit)$coefficients[,1][2]
effect
#effect:0.000319244 
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr =2.864277e-05


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           abdominal_pelvic_rt.ch1 +
                           alkylating.agent.classic.ch1+
                           anthracyclines.ch1 +
                           corticosteroids.ch1 +
                           epipodophyllotoxins.ch1 +
                           platinum.ch1 +
                           vincristine.ch1 +
                           chestrt.ch1 +
                           brainrt.ch1 +
                           CD4T +
                           Bcell +
                           CD8T +
                           NK +
                           Gran,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE197674_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197674 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 1.718455e-44   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect:0.0003351834 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 2.340389e-05   

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.825406    

tiff('GSE197674 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.825406",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE197674 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex +
                        abdominal_pelvic_rt.ch1 +
                        alkylating.agent.classic.ch1+
                        anthracyclines.ch1 +
                        corticosteroids.ch1 +
                        epipodophyllotoxins.ch1 +
                        platinum.ch1 +
                        vincristine.ch1 +
                        chestrt.ch1 +
                        brainrt.ch1,
                      pheno)

fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197674 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197674 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE197674 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 3.177914e-43              
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0005135328          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 3.642573e-05    

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.0005021732               
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 8.187229e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 2.349317e-05    


write.table(pheno, "GSE197674 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)





pheno <- read.delim("GSE197674 Phenotypes.txt")

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE197674 Entropy all age-related CpGs vs non-age-related CpGs 2.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`, colour = Condition))+ 
  geom_jitter(alpha = 0.3, width = 0.6) +
  geom_smooth(method = "lm")+
  xlim(0,80)+
  ylim(0.3,0.85)+
  scale_colour_manual(values = c("#3F7C85","#747E7E"),
                      name = "", 
                      labels = c("All age-related CpGs","Non age-related CpGs"))+
  scale_alpha(guide = 'none')+
  ggtitle("GSE197674")+
  theme_minimal() +
  theme(plot.title = element_text(size = 12,hjust = 0.5))
dev.off()



#### GSE197676 ####

path = "/pvol/Preprocessing/Blood/GEO/GSE197676"
setwd(path)

library(tidyverse)
library(limma)
library(minfi)
library(ChAMP)
library(GEOquery)
setwd('')
#### DMPs 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,l
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE197676 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE197676 Phenotypes.txt")
glimpse(pheno)


#DMPs
design=model.matrix(~age +
                      sex,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE197676_pvalhist_DMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#184544 DMPs 

directory = "/pvol/EWAS/Blood/DMPs"
create_summary(toptable = results,
               dataset_label = "GSE197676",
               directory = directory)

#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE197676_M_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE197676_DMP_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### VMPs
resid <- read.table("GSE197676_M_res.txt")
pheno <- read.delim("GSE197676 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,] 
#353581 cpgs
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]

TestStat["cg08319905",] #check that the columns have aligned correctly

#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)

results_metaRE <- tibble(CPG = rownames(results),
                         TESTSTAT = results$TestStat,
                         N = rep(ncol(w),nrow(w)),
                         COEF = results$logFC,
                         SE = results$SE,
                         PVALUE = results$chisq_pval,
                         BP = ncol(resid)*fitted_sqrd/rowSums(w^2))
#Save p-value histogram
tiff('GSE197676_pvalhist_VMPs.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#6299 VMPs with pval 

setwd("/pvol/EWAS/Blood/VMPs")
write.table(results_BP,
            file="GSE197676.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

setwd("/pvol/EWAS/Blood/RE meta")
write.table(results_metaRE,
            file="GSE197676_VMPs.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE197676"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE197676_VMPcheck.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### DMPs CTC 
create_summary <- function(toptable = NULL,
                           dataset_label = NULL,
                           directory = getwd())
{
  CPG <- rownames(toptable)
  ALLELE1 <- rep(1,nrow(toptable))
  ALLELE2 <- rep(2,nrow(toptable))
  TESTSTAT <- toptable$t
  PVALUE <- toptable$P.Value
  EFFECTSIZE <- toptable$logFC
  SE <- toptable$SE
  results = data.frame(CPG,
                       ALLELE1,
                       ALLELE2,
                       TESTSTAT,
                       PVALUE,
                       EFFECTSIZE,
                       SE)
  write.table(results,
              file=paste0(directory,"/",dataset_label,".tbl"),
              quote = FALSE,
              row.names = FALSE,
              col.names = TRUE,
              sep="\t")
}

#load the data 
B <- data.table::fread("GSE197676 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE197676 Phenotypes.txt")
glimpse(pheno)

celltypes <- champ.refbase(beta = B,
                           arraytype = "EPIC")
celltypes <- celltypes[[2]]
celltypes <- as.data.frame(celltypes)
celltypes$Sample_Name <- rownames(celltypes)
pheno <- left_join(pheno, celltypes, by = "Sample_Name")

write.table(pheno, 
            "GSE197676 Phenotypes.txt",
            col.names = TRUE,
            sep = "\t")

#Use EpiDISH package to obtain cell type estimates
library(EpiDISH)
out.l <- epidish(beta.m = B,
                 ref.m = centDHSbloodDMC.m[,1:6],
                 method = "RPC") 
frac.m <- out.l$estF

#Run cellDMC
results <- CellDMC(beta.m = as.matrix(B),
                   pheno.v = pheno$age,
                   frac.m = frac.m,
                   adjPThresh = 1,
                   cov.mod = model.matrix(~sex,
                                          pheno))

#Save results for all coefficients related to age: Age alone, Age:NK, Age:CD4T,...
saveRDS(results$coe,
        file = '/pvol/prelim_data_ARC_2023/CellDMC_GSE197676.rds')

#DMPs
design=model.matrix(~age +
                      sex +
                      CD4T +
                      CD8T +
                      Mono +
                      NK +
                      Bcell,
                    pheno)

fit1_M <- lmFit(M,
                design)
fit2_M <- eBayes(fit1_M)

fit1_B <- lmFit(B,
                design)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2_M,
                    coef=coef,
                    number = Inf,
                    p.value = 1)
results_B <- topTable(fit2_B,
                      coef=coef,
                      number=Inf,
                      p.value=1)
results$logFC <- results_B[rownames(results),"logFC"]
SE <- fit2_B$sigma * fit2_B$stdev.unscaled
results$SE <- SE[rownames(results),coef]

#Save p-value histogram
tiff('GSE197676_pvalhist_DMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results,
       mapping = aes(x = `P.Value`))+
  labs(title="Distribution of raw p-values for age DMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue",bins = 30)
dev.off()

#for interest check number of DMPs
fdr = 0.005
results_age=topTable(fit2_M,
                     coef = "age",
                     number = nrow(M),
                     adjust.method = "BH",
                     p.value = fdr)
#224469 DMPs 

directory = "/pvol/EWAS/Blood/DMPs with cell type correction"
create_summary(toptable = results,
               dataset_label = "GSE197676",
               directory = directory)

#check to see if limma worked

cpg <-rownames(results)[1]
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE197676_DMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter(width = 0.06)+
  labs(y=paste("Methylation at",cpg))
dev.off()


#Save residuals
resid <- residuals(fit2_M, M)
write.table(signif(resid,digits = 4),
            file="GSE197676_M_CTC_res.txt",
            quote = FALSE,
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")


#### VMPs CTC
resid <- read.table("GSE197676_M_CTC_res.txt")
pheno <- read.delim("GSE197676 Phenotypes.txt")

design2 = model.matrix(~ age, pheno)

#Identify non-normal CpGs and remove them
shapirotest <- apply(as.matrix(resid),1,shapiro.test)
pvals <- sapply(shapirotest,"[[",2)
CpGs_to_keep <- names(pvals[pvals>1e-5])
#Remove all sites with raw p-value < 1e-5 (edited) 
resid <- resid[CpGs_to_keep,]
#299444 cpgs in residuals
resid_B <- ilogit2(resid)

#regression of M squared residuals 
sigma2 <- rowSums(resid^2)/ncol(resid) #variance of the residuals
w <- (resid^2) - sigma2
w <- as.matrix(w)

fit1 <- lmFit(w,
              design2)
fit2 <- eBayes(fit1)

#run second regression with beta values to get the effect size and standard error in beta values
sigma2_B <- rowSums(resid_B^2)/ncol(resid_B) #variance of the residuals
w_B <- (resid_B^2) - sigma2_B
w_B <- as.matrix(w_B)

fit1_B <- lmFit(w_B,
                design2)
fit2_B <- eBayes(fit1_B)

coef = "age"
results <- topTable(fit2,
                    coef=coef,
                    number=Inf,
                    p.value=1)
results_B <- topTable(fit2_B,
                      coef = coef,
                      number = Inf, 
                      p.value = 1)

results$logFC <- results_B[rownames(results),"logFC"] #extract logFC in beta values
SE <- fit2_B$sigma * fit2_B$stdev.unscaled #extract SE in beta values
results$SE <- SE[rownames(results),coef]

fitted <- fitted(fit2)
fitted_sqrd <- rowSums(fitted^2)
bp <- ncol(resid)*fitted_sqrd/rowSums(w^2)
df <- fit2$rank -1
chisq_pval = sapply(bp,pchisq,df=df,lower.tail = FALSE)
chisq_pval <- as.data.frame(chisq_pval)
results$chisq_pval <- chisq_pval[rownames(results),]
TestStat <- as.data.frame(bp)
names(TestStat)[names(TestStat) == "bp"] <- "TestStat"
results$TestStat <- TestStat[rownames(results), ]


#format file for METAL
results_BP <- tibble(CPG = rownames(results),
                     ALLELE1 = rep(1,nrow(results)),
                     ALLELE2 = rep(2,nrow(results)),
                     TESTSTAT = results$TestStat,
                     N = rep(ncol(w),nrow(w)),
                     COEF = results$logFC,
                     SE = results$SE,
                     PVALUE = results$chisq_pval)
#Save p-value histogram
tiff('GSE197676_pvalhist_VMPsCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(results_BP,
       mapping = aes(x = PVALUE, binwidth = 20))+
  labs(title="Distribution of raw p-values for age VMPs with CTC",
       x="p-value")+
  geom_histogram(color="darkblue",
                 fill="lightblue")
dev.off()

filter(results_BP, PVALUE < 0.005)
#3907 VMPs 

setwd("/pvol/EWAS/Blood/VMPs with cell type correction")
write.table(results_BP,
            file="GSE197676.tbl",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE,
            sep="\t")

path = "/pvol/Preprocessing/Blood/GEO/GSE197676"
setwd(path)
#VMP check
cpg <- as.character(results_BP[1,1])
pheno_with_meth <- cbind(pheno, meth= as.numeric(B[cpg,]))
tiff('GSE197676_VMP_CTC_check.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno_with_meth, aes(x=age, y=meth)) +
  geom_jitter()+
  labs(y=paste("Methylation at",cpg))
dev.off()

#### Entropy 

B <- data.table::fread("GSE197676 beta after normalisation and batch correction.txt")
B <- as.data.frame(B)
rownames(B) <- B$V1
B <- B %>% select(-V1)
M <- logit2(B)
pheno <- read.delim("GSE197676 Phenotypes.txt")

#run limma and adjust for confounders EXCLUDING age
design = model.matrix(~sex,
                      pheno)

fit1 <- lmFit(M,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy <- apply(B_adj, 2, calculate_entropy)
Entropy <- as.data.frame(Entropy)
Entropy$Sample_Name <- rownames(Entropy)
pheno <- left_join(pheno, Entropy, by = "Sample_Name")

#Save entropy plot
tiff('GSE197676_Entropy.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197676 Entropy",
       ylab = "entropy")+
  theme_minimal() 
dev.off()


fit <- lm(Entropy ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#1.646378e-07   
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0003769866    
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr = 7.018725e-05 


#Entropy with CTC 

designCTC = model.matrix(~sex +
                           CD4T +
                           CD8T +
                           Mono +
                           NK +
                           Bcell,
                         pheno)
fit1_CTC <- lmFit(M,
                  designCTC)
fit2_CTC <- eBayes(fit1_CTC)

resid_M_CTC <- residuals(fit2_CTC, M)
M_CTC_adj <- resid_M_CTC + rowMeans(M)
B_CTC_adj <- ilogit2(M_CTC_adj)
plotDensities(B_CTC_adj, legend = F)

Entropy_CTC <- apply(B_CTC_adj, 2, calculate_entropy)
Entropy_CTC <- as.data.frame(Entropy_CTC)
Entropy_CTC$Sample_Name<- rownames(Entropy_CTC)
pheno <- left_join(pheno, Entropy_CTC, by = "Sample_Name")

#Save entropy plot
tiff('GSE197676_EntropyCTC.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(data = pheno, aes(x = age, y = Entropy_CTC)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197676 Entropy with cell type correction",
       ylab = "entropy")+
  theme_minimal()
dev.off()


fitCTC <- lm(Entropy_CTC ~ age, pheno)
summary(fitCTC)
pvalCTC = summary(fitCTC)$coefficients[,4][2]
pvalCTC
#p-value: 4.110603e-09   
effectCTC = summary(fitCTC)$coefficients[,1][2]
effectCTC
#effect: 0.0003797449 
stderrCTC = summary(fitCTC)$coefficients[,2][2]
stderrCTC
#stderr = 6.254422e-05  

cor.test(pheno$Entropy, pheno$Entropy_CTC, method = "pearson")
#pval =   < 2.2e-16
#cor = 0.8943991   

tiff('GSE197676 Entropy ~ Entropy with cell type correction.tiff',
     width =5,
     height = 3,
     units = 'in',
     res = 200)
ggplot(pheno) +
  geom_jitter(aes(x = Entropy, y = Entropy_CTC), colour = "cornflowerblue") +
  geom_smooth(method = "lm", aes(x = Entropy, y = Entropy_CTC), colour = "black") +
  labs(title = "Pearson's corr = 0.8943991",
       ylabs = "Entropy with cell type correction")+
  theme_minimal()
dev.off()

write.table(pheno, "GSE197676 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)

#Entropy on all age-related cpgs
DMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/DMPs/Blood_DMPs.csv")
VMPs <- read.csv("/pvol/EWAS/Blood/Meta-analysis/VMPs/Blood_VMPs.csv")
All_cps <- full_join(DMPs, VMPs, by = "MarkerName")
All_cps <- All_cps$MarkerName

B <- as.data.frame(B)
B_all <- B[rownames(B) %in% All_cps,]
M_all <- logit2(B_all)

#run limma and adjust for confounders EXCLUDING age

design = model.matrix(~sex,
                      pheno)
fit1 <- lmFit(M_all,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_all)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_all)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_all <- apply(B_adj, 2, calculate_entropy)
Entropy_all <- as.data.frame(Entropy_all)
Entropy_all$Sample_Name <- rownames(Entropy_all)
pheno$Sample_Name <- as.character(pheno$Sample_Name)
pheno <- left_join(pheno, Entropy_all, by = "Sample_Name")


ggplot(data = pheno, aes(x = age, y = Entropy_all)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197676 Entropy all age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

#calculate entropy on cpgs that do not change with age 

B_none <- B[!rownames(B) %in% All_cps,]
M_none <- logit2(B_none)

fit1 <- lmFit(M_none,
              design)
fit2 <- eBayes(fit1)

resid_M <- residuals(fit2, M_none)

#add the mean beta value for each cpg to the residuals to get a beta matrix 

M_adj <- resid_M + rowMeans(M_none)
B_adj <- ilogit2(M_adj)
plotDensities(B_adj, legend = F)

#calcuate entropy 
calculate_entropy <- function(B){
  return(sum(B*log2(B) + (1-B)*log2(1-B))/(length(B)*log2(1/2)))
}

Entropy_none <- apply(B_adj, 2, calculate_entropy)
Entropy_none <- as.data.frame(Entropy_none)
Entropy_none$Sample_Name <- rownames(Entropy_none)
pheno <- left_join(pheno, Entropy_none, by = "Sample_Name")

ggplot(data = pheno, aes(x = age, y = Entropy_none)) +
  geom_jitter(colour = "deepskyblue4") +
  geom_smooth(method = "lm", colour = "black") +
  labs(title = "GSE197676 Entropy all non-age-related CpGs",
       ylab = "entropy")+
  theme_minimal()

pheno_entropy <- pheno %>% select(Sample_Name, age, Entropy_all, Entropy_none)

pheno_entropy <- pheno_entropy %>% pivot_longer(!c(age,Sample_Name),
                                                names_to = "Entropy type",
                                                values_to = "Entropy value")

names(pheno_entropy)[names(pheno_entropy) == "Entropy type"] <- "Condition"

library(viridis)
tiff('GSE197676 Entropy all age-related CpGs vs non-age-related CpGs.tiff',
     width =10,
     height = 5,
     units = 'in',
     res = 400)
#pheno_entropy <- pheno_entropy %>% mutate(Class = factor(Class, levels = c("Entropy_VMPs","Entropy_DMPs","Entropy")))
ggplot(pheno_entropy, aes(x = age, y = `Entropy value`))+ 
  geom_jitter(aes(colour = Condition, alpha = 0.8)) +
  geom_smooth(aes(colour = Condition), method = "lm")+ 
  scale_color_viridis(discrete = TRUE, option = "D")+
  scale_fill_viridis(discrete = TRUE) +
  theme_minimal()
dev.off()


#summary statistics all 
fit <- lm(Entropy_all ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 3.320508e-12                
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 0.0006357753          
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 8.729504e-05      

#summary statistics non-age-related cpgs
fit <- lm(Entropy_none ~ age, pheno)
summary(fit)
pval = summary(fit)$coefficients[,4][2]
pval
#p-value: 0.4128983                  
effect = summary(fit)$coefficients[,1][2]
effect
#effect: 5.809466e-05        
stderr = summary(fit)$coefficients[,2][2]
stderr
#stderr: 7.084483e-05      


write.table(pheno, "GSE197676 Phenotypes.txt",
            sep = "\t",
            col.names = TRUE)



